<!DOCTYPE html>
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>

<title>Introduction</title>

<script type="text/javascript">
window.onload = function() {
  var imgs = document.getElementsByTagName('img'), i, img;
  for (i = 0; i < imgs.length; i++) {
    img = imgs[i];
    // center an image if it is the only element of its parent
    if (img.parentElement.childElementCount === 1)
      img.parentElement.style.textAlign = 'center';
  }
};
</script>

<!-- Styles for R syntax highlighter -->
<style type="text/css">
   pre .operator,
   pre .paren {
     color: rgb(104, 118, 135)
   }

   pre .literal {
     color: #990073
   }

   pre .number {
     color: #099;
   }

   pre .comment {
     color: #998;
     font-style: italic
   }

   pre .keyword {
     color: #900;
     font-weight: bold
   }

   pre .identifier {
     color: rgb(0, 0, 0);
   }

   pre .string {
     color: #d14;
   }
</style>

<!-- R syntax highlighter -->
<script type="text/javascript">
var hljs=new function(){function m(p){return p.replace(/&/gm,"&amp;").replace(/</gm,"&lt;")}function f(r,q,p){return RegExp(q,"m"+(r.cI?"i":"")+(p?"g":""))}function b(r){for(var p=0;p<r.childNodes.length;p++){var q=r.childNodes[p];if(q.nodeName=="CODE"){return q}if(!(q.nodeType==3&&q.nodeValue.match(/\s+/))){break}}}function h(t,s){var p="";for(var r=0;r<t.childNodes.length;r++){if(t.childNodes[r].nodeType==3){var q=t.childNodes[r].nodeValue;if(s){q=q.replace(/\n/g,"")}p+=q}else{if(t.childNodes[r].nodeName=="BR"){p+="\n"}else{p+=h(t.childNodes[r])}}}if(/MSIE [678]/.test(navigator.userAgent)){p=p.replace(/\r/g,"\n")}return p}function a(s){var r=s.className.split(/\s+/);r=r.concat(s.parentNode.className.split(/\s+/));for(var q=0;q<r.length;q++){var p=r[q].replace(/^language-/,"");if(e[p]){return p}}}function c(q){var p=[];(function(s,t){for(var r=0;r<s.childNodes.length;r++){if(s.childNodes[r].nodeType==3){t+=s.childNodes[r].nodeValue.length}else{if(s.childNodes[r].nodeName=="BR"){t+=1}else{if(s.childNodes[r].nodeType==1){p.push({event:"start",offset:t,node:s.childNodes[r]});t=arguments.callee(s.childNodes[r],t);p.push({event:"stop",offset:t,node:s.childNodes[r]})}}}}return t})(q,0);return p}function k(y,w,x){var q=0;var z="";var s=[];function u(){if(y.length&&w.length){if(y[0].offset!=w[0].offset){return(y[0].offset<w[0].offset)?y:w}else{return w[0].event=="start"?y:w}}else{return y.length?y:w}}function t(D){var A="<"+D.nodeName.toLowerCase();for(var B=0;B<D.attributes.length;B++){var C=D.attributes[B];A+=" "+C.nodeName.toLowerCase();if(C.value!==undefined&&C.value!==false&&C.value!==null){A+='="'+m(C.value)+'"'}}return A+">"}while(y.length||w.length){var v=u().splice(0,1)[0];z+=m(x.substr(q,v.offset-q));q=v.offset;if(v.event=="start"){z+=t(v.node);s.push(v.node)}else{if(v.event=="stop"){var p,r=s.length;do{r--;p=s[r];z+=("</"+p.nodeName.toLowerCase()+">")}while(p!=v.node);s.splice(r,1);while(r<s.length){z+=t(s[r]);r++}}}}return z+m(x.substr(q))}function j(){function q(x,y,v){if(x.compiled){return}var u;var s=[];if(x.k){x.lR=f(y,x.l||hljs.IR,true);for(var w in x.k){if(!x.k.hasOwnProperty(w)){continue}if(x.k[w] instanceof Object){u=x.k[w]}else{u=x.k;w="keyword"}for(var r in u){if(!u.hasOwnProperty(r)){continue}x.k[r]=[w,u[r]];s.push(r)}}}if(!v){if(x.bWK){x.b="\\b("+s.join("|")+")\\s"}x.bR=f(y,x.b?x.b:"\\B|\\b");if(!x.e&&!x.eW){x.e="\\B|\\b"}if(x.e){x.eR=f(y,x.e)}}if(x.i){x.iR=f(y,x.i)}if(x.r===undefined){x.r=1}if(!x.c){x.c=[]}x.compiled=true;for(var t=0;t<x.c.length;t++){if(x.c[t]=="self"){x.c[t]=x}q(x.c[t],y,false)}if(x.starts){q(x.starts,y,false)}}for(var p in e){if(!e.hasOwnProperty(p)){continue}q(e[p].dM,e[p],true)}}function d(B,C){if(!j.called){j();j.called=true}function q(r,M){for(var L=0;L<M.c.length;L++){if((M.c[L].bR.exec(r)||[null])[0]==r){return M.c[L]}}}function v(L,r){if(D[L].e&&D[L].eR.test(r)){return 1}if(D[L].eW){var M=v(L-1,r);return M?M+1:0}return 0}function w(r,L){return L.i&&L.iR.test(r)}function K(N,O){var M=[];for(var L=0;L<N.c.length;L++){M.push(N.c[L].b)}var r=D.length-1;do{if(D[r].e){M.push(D[r].e)}r--}while(D[r+1].eW);if(N.i){M.push(N.i)}return f(O,M.join("|"),true)}function p(M,L){var N=D[D.length-1];if(!N.t){N.t=K(N,E)}N.t.lastIndex=L;var r=N.t.exec(M);return r?[M.substr(L,r.index-L),r[0],false]:[M.substr(L),"",true]}function z(N,r){var L=E.cI?r[0].toLowerCase():r[0];var M=N.k[L];if(M&&M instanceof Array){return M}return false}function F(L,P){L=m(L);if(!P.k){return L}var r="";var O=0;P.lR.lastIndex=0;var M=P.lR.exec(L);while(M){r+=L.substr(O,M.index-O);var N=z(P,M);if(N){x+=N[1];r+='<span class="'+N[0]+'">'+M[0]+"</span>"}else{r+=M[0]}O=P.lR.lastIndex;M=P.lR.exec(L)}return r+L.substr(O,L.length-O)}function J(L,M){if(M.sL&&e[M.sL]){var r=d(M.sL,L);x+=r.keyword_count;return r.value}else{return F(L,M)}}function I(M,r){var L=M.cN?'<span class="'+M.cN+'">':"";if(M.rB){y+=L;M.buffer=""}else{if(M.eB){y+=m(r)+L;M.buffer=""}else{y+=L;M.buffer=r}}D.push(M);A+=M.r}function G(N,M,Q){var R=D[D.length-1];if(Q){y+=J(R.buffer+N,R);return false}var P=q(M,R);if(P){y+=J(R.buffer+N,R);I(P,M);return P.rB}var L=v(D.length-1,M);if(L){var O=R.cN?"</span>":"";if(R.rE){y+=J(R.buffer+N,R)+O}else{if(R.eE){y+=J(R.buffer+N,R)+O+m(M)}else{y+=J(R.buffer+N+M,R)+O}}while(L>1){O=D[D.length-2].cN?"</span>":"";y+=O;L--;D.length--}var r=D[D.length-1];D.length--;D[D.length-1].buffer="";if(r.starts){I(r.starts,"")}return R.rE}if(w(M,R)){throw"Illegal"}}var E=e[B];var D=[E.dM];var A=0;var x=0;var y="";try{var s,u=0;E.dM.buffer="";do{s=p(C,u);var t=G(s[0],s[1],s[2]);u+=s[0].length;if(!t){u+=s[1].length}}while(!s[2]);if(D.length>1){throw"Illegal"}return{r:A,keyword_count:x,value:y}}catch(H){if(H=="Illegal"){return{r:0,keyword_count:0,value:m(C)}}else{throw H}}}function g(t){var p={keyword_count:0,r:0,value:m(t)};var r=p;for(var q in e){if(!e.hasOwnProperty(q)){continue}var s=d(q,t);s.language=q;if(s.keyword_count+s.r>r.keyword_count+r.r){r=s}if(s.keyword_count+s.r>p.keyword_count+p.r){r=p;p=s}}if(r.language){p.second_best=r}return p}function i(r,q,p){if(q){r=r.replace(/^((<[^>]+>|\t)+)/gm,function(t,w,v,u){return w.replace(/\t/g,q)})}if(p){r=r.replace(/\n/g,"<br>")}return r}function n(t,w,r){var x=h(t,r);var v=a(t);var y,s;if(v){y=d(v,x)}else{return}var q=c(t);if(q.length){s=document.createElement("pre");s.innerHTML=y.value;y.value=k(q,c(s),x)}y.value=i(y.value,w,r);var u=t.className;if(!u.match("(\\s|^)(language-)?"+v+"(\\s|$)")){u=u?(u+" "+v):v}if(/MSIE [678]/.test(navigator.userAgent)&&t.tagName=="CODE"&&t.parentNode.tagName=="PRE"){s=t.parentNode;var p=document.createElement("div");p.innerHTML="<pre><code>"+y.value+"</code></pre>";t=p.firstChild.firstChild;p.firstChild.cN=s.cN;s.parentNode.replaceChild(p.firstChild,s)}else{t.innerHTML=y.value}t.className=u;t.result={language:v,kw:y.keyword_count,re:y.r};if(y.second_best){t.second_best={language:y.second_best.language,kw:y.second_best.keyword_count,re:y.second_best.r}}}function o(){if(o.called){return}o.called=true;var r=document.getElementsByTagName("pre");for(var p=0;p<r.length;p++){var q=b(r[p]);if(q){n(q,hljs.tabReplace)}}}function l(){if(window.addEventListener){window.addEventListener("DOMContentLoaded",o,false);window.addEventListener("load",o,false)}else{if(window.attachEvent){window.attachEvent("onload",o)}else{window.onload=o}}}var e={};this.LANGUAGES=e;this.highlight=d;this.highlightAuto=g;this.fixMarkup=i;this.highlightBlock=n;this.initHighlighting=o;this.initHighlightingOnLoad=l;this.IR="[a-zA-Z][a-zA-Z0-9_]*";this.UIR="[a-zA-Z_][a-zA-Z0-9_]*";this.NR="\\b\\d+(\\.\\d+)?";this.CNR="\\b(0[xX][a-fA-F0-9]+|(\\d+(\\.\\d*)?|\\.\\d+)([eE][-+]?\\d+)?)";this.BNR="\\b(0b[01]+)";this.RSR="!|!=|!==|%|%=|&|&&|&=|\\*|\\*=|\\+|\\+=|,|\\.|-|-=|/|/=|:|;|<|<<|<<=|<=|=|==|===|>|>=|>>|>>=|>>>|>>>=|\\?|\\[|\\{|\\(|\\^|\\^=|\\||\\|=|\\|\\||~";this.ER="(?![\\s\\S])";this.BE={b:"\\\\.",r:0};this.ASM={cN:"string",b:"'",e:"'",i:"\\n",c:[this.BE],r:0};this.QSM={cN:"string",b:'"',e:'"',i:"\\n",c:[this.BE],r:0};this.CLCM={cN:"comment",b:"//",e:"$"};this.CBLCLM={cN:"comment",b:"/\\*",e:"\\*/"};this.HCM={cN:"comment",b:"#",e:"$"};this.NM={cN:"number",b:this.NR,r:0};this.CNM={cN:"number",b:this.CNR,r:0};this.BNM={cN:"number",b:this.BNR,r:0};this.inherit=function(r,s){var p={};for(var q in r){p[q]=r[q]}if(s){for(var q in s){p[q]=s[q]}}return p}}();hljs.LANGUAGES.cpp=function(){var a={keyword:{"false":1,"int":1,"float":1,"while":1,"private":1,"char":1,"catch":1,"export":1,virtual:1,operator:2,sizeof:2,dynamic_cast:2,typedef:2,const_cast:2,"const":1,struct:1,"for":1,static_cast:2,union:1,namespace:1,unsigned:1,"long":1,"throw":1,"volatile":2,"static":1,"protected":1,bool:1,template:1,mutable:1,"if":1,"public":1,friend:2,"do":1,"return":1,"goto":1,auto:1,"void":2,"enum":1,"else":1,"break":1,"new":1,extern:1,using:1,"true":1,"class":1,asm:1,"case":1,typeid:1,"short":1,reinterpret_cast:2,"default":1,"double":1,register:1,explicit:1,signed:1,typename:1,"try":1,"this":1,"switch":1,"continue":1,wchar_t:1,inline:1,"delete":1,alignof:1,char16_t:1,char32_t:1,constexpr:1,decltype:1,noexcept:1,nullptr:1,static_assert:1,thread_local:1,restrict:1,_Bool:1,complex:1},built_in:{std:1,string:1,cin:1,cout:1,cerr:1,clog:1,stringstream:1,istringstream:1,ostringstream:1,auto_ptr:1,deque:1,list:1,queue:1,stack:1,vector:1,map:1,set:1,bitset:1,multiset:1,multimap:1,unordered_set:1,unordered_map:1,unordered_multiset:1,unordered_multimap:1,array:1,shared_ptr:1}};return{dM:{k:a,i:"</",c:[hljs.CLCM,hljs.CBLCLM,hljs.QSM,{cN:"string",b:"'\\\\?.",e:"'",i:"."},{cN:"number",b:"\\b(\\d+(\\.\\d*)?|\\.\\d+)(u|U|l|L|ul|UL|f|F)"},hljs.CNM,{cN:"preprocessor",b:"#",e:"$"},{cN:"stl_container",b:"\\b(deque|list|queue|stack|vector|map|set|bitset|multiset|multimap|unordered_map|unordered_set|unordered_multiset|unordered_multimap|array)\\s*<",e:">",k:a,r:10,c:["self"]}]}}}();hljs.LANGUAGES.r={dM:{c:[hljs.HCM,{cN:"number",b:"\\b0[xX][0-9a-fA-F]+[Li]?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+(?:[eE][+\\-]?\\d*)?L\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+\\.(?!\\d)(?:i\\b)?",e:hljs.IMMEDIATE_RE,r:1},{cN:"number",b:"\\b\\d+(?:\\.\\d*)?(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\.\\d+(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"keyword",b:"(?:tryCatch|library|setGeneric|setGroupGeneric)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\.",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\d+(?![\\w.])",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\b(?:function)",e:hljs.IMMEDIATE_RE,r:2},{cN:"keyword",b:"(?:if|in|break|next|repeat|else|for|return|switch|while|try|stop|warning|require|attach|detach|source|setMethod|setClass)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"literal",b:"(?:NA|NA_integer_|NA_real_|NA_character_|NA_complex_)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"literal",b:"(?:NULL|TRUE|FALSE|T|F|Inf|NaN)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"identifier",b:"[a-zA-Z.][a-zA-Z0-9._]*\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"<\\-(?!\\s*\\d)",e:hljs.IMMEDIATE_RE,r:2},{cN:"operator",b:"\\->|<\\-",e:hljs.IMMEDIATE_RE,r:1},{cN:"operator",b:"%%|~",e:hljs.IMMEDIATE_RE},{cN:"operator",b:">=|<=|==|!=|\\|\\||&&|=|\\+|\\-|\\*|/|\\^|>|<|!|&|\\||\\$|:",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"%",e:"%",i:"\\n",r:1},{cN:"identifier",b:"`",e:"`",r:0},{cN:"string",b:'"',e:'"',c:[hljs.BE],r:0},{cN:"string",b:"'",e:"'",c:[hljs.BE],r:0},{cN:"paren",b:"[[({\\])}]",e:hljs.IMMEDIATE_RE,r:0}]}};
hljs.initHighlightingOnLoad();
</script>

<!-- MathJax scripts -->
<script type="text/javascript" src="https://cdn.bootcss.com/mathjax/2.7.0/MathJax.js?config=TeX-MML-AM_CHTML">
</script>


<style type="text/css">
body, td {
   font-family: sans-serif;
   background-color: white;
   font-size: 13px;
}

body {
  max-width: 800px;
  margin: auto;
  padding: 1em;
  line-height: 20px;
}

tt, code, pre {
   font-family: 'DejaVu Sans Mono', 'Droid Sans Mono', 'Lucida Console', Consolas, Monaco, monospace;
}

h1 {
   font-size:2.2em;
}

h2 {
   font-size:1.8em;
}

h3 {
   font-size:1.4em;
}

h4 {
   font-size:1.0em;
}

h5 {
   font-size:0.9em;
}

h6 {
   font-size:0.8em;
}

a:visited {
   color: rgb(50%, 0%, 50%);
}

pre, img {
  max-width: 100%;
}
pre {
  overflow-x: auto;
}
pre code {
   display: block; padding: 0.5em;
}

code {
  font-size: 92%;
  border: 1px solid #ccc;
}

code[class] {
  background-color: #F8F8F8;
}

table, td, th {
  border: none;
}

blockquote {
   color:#666666;
   margin:0;
   padding-left: 1em;
   border-left: 0.5em #EEE solid;
}

hr {
   height: 0px;
   border-bottom: none;
   border-top-width: thin;
   border-top-style: dotted;
   border-top-color: #999999;
}

@media print {
   * {
      background: transparent !important;
      color: black !important;
      filter:none !important;
      -ms-filter: none !important;
   }

   body {
      font-size:12pt;
      max-width:100%;
   }

   a, a:visited {
      text-decoration: underline;
   }

   hr {
      visibility: hidden;
      page-break-before: always;
   }

   pre, blockquote {
      padding-right: 1em;
      page-break-inside: avoid;
   }

   tr, img {
      page-break-inside: avoid;
   }

   img {
      max-width: 100% !important;
   }

   @page :left {
      margin: 15mm 20mm 15mm 10mm;
   }

   @page :right {
      margin: 15mm 10mm 15mm 20mm;
   }

   p, h2, h3 {
      orphans: 3; widows: 3;
   }

   h2, h3 {
      page-break-after: avoid;
   }
}
</style>



</head>

<body>
<p>\abstract{
Subgroup discovery is a data mining task half-way between descriptive and predictive data mining. Nowadays it is 
very relevant for researches due to the fact that the knowledge extracted is simple and interesting.
For this task, evolutionary fuzzy systems are well suited algorithms because they can find a good 
trade-off between multiple objectives in large search spaces. In fact, this paper presents the SDEFSR
package, which contains all evolutionary fuzzy systems for subgroup discovery 
presented throughout the literature. It is a package without dependencies on other software, providing
functions with recommended default parameters. In addition, it brings a graphical user interface to avoid the user having to know all the algorithm&#39;s parameters.
}</p>

<h1>Introduction</h1>

<p>Subgroup discovery (SD) is a data mining field that aims to describe data using supervised learning techniques. The goal is to find simple, general and interesting patterns with respect to a given variable of interest. Throughout the literature, SD has been applied with success to different real-world problems in areas such as marketing \cite{deljesus2007evolutionary, mesdif}, medicine \cite{carmona2015fuzzy, carmona2013fuzzy, Stiglic2012348, Gamberger200327} and e-learning \cite{Poitras2016204, Lemmerich2011129, carmona2010evolutionary}, amongst others \cite{Atzmueller2016965,  Jin2014SmartElectricity, Rodriguez20131810, OrOlive}.   </p>

<p>SD is a rule learning process within a complex search space. Therefore, the search strategy used becomes a key factor in the efficiency of the method. Different strategies can be found in the literature such as beam search in the algorithm CN2-SD \cite{lavravc2004subgroup} and Apriori-SD \cite{kavvsek2006apriori}, exhaustive algorithms such as SDMap \cite{atzmueller2006sd} or Evolutionary Algorithms (EAs), for example.</p>

<p>EAs are stochastic algorithms for optimizing and searching tasks, based on the natural evolution process \cite{john1992holland}. There are different paradigms within EAs: genetic algorithms \cite{john1992holland, goldberg1989genetic}, evolution strategies \cite{schwefel1995evolution}, evolutionary programming \cite{fogel2006evolutionary} and genetic programming \cite{koza1992genetic}. With these methods the inclusion of rules as knowledge representation is known as evolutionary rule-based systems \cite{freitas2003survey} and has the advantage of allowing the inclusion of domain knowledge and returning better rules. The use of EAs is very well suited to SD because these algorithms perform a global search in the space in a convenient way, stimulating the obtaining of simple, interesting and precise subgroups.</p>

<p>Nowadays, there are several frameworks that allow the realization of data mining tasks, but only a few of them have implementations of SD algorithms. The most well-know frameworks with SD algorithms are KEEL \cite{keelDatasets} and VIKAMINE \cite{atzmueller2012vikamine}, but ORANGE \cite{demvsar2013orange}, and CORTANA \cite{meeng2011flexible} provide some implementations too. In fact, VIKAMINE also provides an R package called <strong>rsubgroup</strong> \cite{rsubgroup} which is an interface for R to VIKAMINE Java algorithms.  </p>

<p>In this contribution the <strong>SDEFSR</strong> package is introduced. It provides the user with the most important evolutionary fuzzy rule-based methods for SD documented in the literature, being a major contribution to the R community. In addition, it also brings the capability of reading datasets in the KEEL data format. This file format is not natively supported by R. Similarly, this package provides a Graphical User Interface (GUI) to make this task easier for the user, especially the unexperienced one.</p>

<p>This contribution is organized according to the following structure: The first section presents SD concepts, main properties and features. In the second section, the structure of the SDEFSR package and its operations are described. In the third section, a complete example of use of the package is presented. Finally, the GUI of SDEFSR is shown.</p>

<h1>Subgroup Discovery</h1>

<p>SD was defined by \cite{definicionSD} as:</p>

<blockquote>
<p>In subgroup discovery, we assume we are given a so-called population of individuals
(objects, customer, &hellip;) and a property of those individuals we are interested in. The
task of subgroup discovery is then to discover the subgroups of the population that are
statistically &ldquo;most interesting&rdquo;, i.e. are as large as possible and have the most unusual
statistical (distributional) characteristics with respect to the property of interest.  </p>
</blockquote>

<p>SD tries to find relations between different properties
of a set with respect to one interesting or target variable. Such relations must be
statistically interesting, so it is not necessary to find complete
relations, partial relations could be interesting too.  </p>

<p>Usually these relations are represented by means of rules, one per relation. These rules are defined as \cite{lavravc2004decision, gamberger2002expert}:
\begin{equation}
R: Cond \rightarrow Target<em>{value}
\end{equation}
where $Target</em>{value}$ is the value for the variable of interest (target
variable) for the SD task and \(Cond\) is normally a
conjunction of attribute-value pairs which describe the characteristics
of the induced subgroup.\
SD is halfway between description and classification, because it has a target
variable but its objective is not to predict it but rather to describe it. The use
of a target variable is not possible in description, because description
simply finds relationships between unlabeled object.\
A key point to fully understanding the goal of SD is how it differentiates from the classification objective.
Classification is a predictive task that tries to split the entire search space, usually in a complex way, aiming to find the correct value for the variable in new incoming instances. On the other hand, SD aims to find interesting relations among these instances regarding the variable of interest.\ For
instance, assuming there is a group of patients, the variable of interest is whether
they have heart disease or not. The predictive data mining
objective is to predict if new patients will have heart disease. However, SD tries to find which
subgroup of patients are more likely to have heart disease according to certain characteristics. This is relevant for developing treatment against those characteristics, for instance.</p>

<h2>Main elements of subgroup discovery algorithms</h2>

<p>Below, the most relevant aspects of SD algorithms are presented \cite{atzmueller2004towards}: 
\begin{itemize}</p>

<pre><code>\item \emph{Type of target variable:} This is the kind of information the interest variable can hold. The target variable could be binary (two possible values), categorical ($n$ posible values) or numerical (a real value within a range). Nevertheless, the majority of SD algorithms can only deal with binary or categorical target variables. 

\item \emph{Description language:} Knowledge representation is a key factor in SD due to its descriptive nature. In this way, rules must be as simple as possible. Rules are usually represented by conjunctions of attribute-value pairs or in disjunctive normal form. Fuzzy logic could also be included in the rules in order to improve the interpretability of the knowledge \cite{zadeh1975concept,hullermeier2005fuzzy}.

\item \emph{Quality measures:} This is the most important aspect in the design of SD algorithms. The quality measures must guide the learning process and must show the quality of the extracted knowledge. They will be briefly described below.

\item \emph{Search strategy:} The search space grows exponentially with the number of variables.The use of a search strategy able to find a good solution, or the optimal one, by searching efficiently through the whole search space is very important.
</code></pre>

<p>\end{itemize}</p>

<h2>Quality measures for subgroup discovery \label{medidas}</h2>

<p>A quality measure tries to measure the interestingness of a given rule or subgroup, but there is not a formal definition of what interestingness is. However, the interestingness could be defined as a concept that emphasizes conciseness, coverage, reliability,
peculiarity, diversity, novelty, surprisingness, utility, and actionability \cite{geng2006interestingness}. For SD, the most used criteria to measure the interestingness of a rule are conciseness, generality or coverage, reliability, and novelty \cite{Carmona2014overview}. </p>

<p>Quality measures that accomplish this criteria available in the <strong>SDEFSR</strong> package are:</p>

<p>\begin{itemize}
    \item Measures for conciseness (or complexity). It measures the complexity of the induced rules. Rules with a few number of attribute-value pairs are easy to remember and to add to the expert&#39;s knowledge. The quality measures associated to this criterion are:
    \begin{itemize}
            \item \(N_r\): The number of rules generated. A rule set with a large number of rules is much more difficult to remember than other that has less rules. Additionally, the lower the number of rules, the easier for the expert to filter those rules that are interesting. 
            \item \(N_v\): The number of variables that have the rules generated. Rules with less number of variables are easier to understand and to remember, also, it tends to have more generality. Thus, rules with low number of variables are interesting.
    \end{itemize}</p>

<pre><code>\item Measures for generality (or coverage). It measures the capacity of a rule to match with a great number of examples in the dataset. Also, the capacity to generalize the rule to other instances that are not in the training dataset is greater. The quality measures associated to this criterion are:
\begin{itemize}
        \item \emph{Support:} It measures the frequency of correctly classified examples covered by the rule: 
        \begin{equation} 
        Sup(R)= \frac{n(Cond \wedge  Target_{value})}{n_s} \label{support}
        \end{equation}
        where $n(Cond \wedge  Target_{value})$ means the number of examples that satisfy the antecedent and consecuent part of the rule and $n_s$ is the number of examples in the dataset. 

        \item \emph{Coverage:} It measures the percentage of examples covered by the rule related to the total number of examples: 
        \begin{equation}
        Cov(R)= \frac{n(Cond)}{n_s} \label{coverage}
        \end{equation} 
        where $n(Cond)$ means the number of examples that satisfy the antecedent part of the rule.
\end{itemize}

\item Measures for reliability. A rule is reliable when the relation described in the rule occurs in a high percentage of cases where the rule can be applied. The quality measures associated to this criterion are:
\begin{itemize}
        \item \emph{Confidence:} It measure the percentage of examples correctly covered of the total of covered examples: 
        \begin{equation}
        Conf(R) = \frac{n(Cond  \wedge Target_{value})}{n(Cond)} \label{confidence} 
        \end{equation}
\end{itemize}

\item Measures for novelty. A rule is novel if the knowledge obtained from this one is unknown by the user or it is unable to infer such knowledge from other rules. For this kind of criterion, the quality measures availables in the package are: 
\begin{itemize}
        \item \emph{Significance:} It reflects the novelty in the distribution of the examples covered by the rule regarding the whole dataset:
        \begin{equation} 
        Sign(R) = 2 \cdot \sum_{k=1}^{n_c}n\left(Cond \wedge Target_{value_k}\right) \cdot log\left( \frac{n\left(Cond \wedge Target_{value_k}\right)}{n(Cond \wedge Target_{value}) \cdot p(Cond)} \right) \label{Significance} 
        \end{equation}
        where $p(Cond) = \frac{n(Cond)}{n_s}$, $n_c$ is the number of possible values of the target variable and $Target_{value_k}$ is the value $k$-th value of the target variable.

\end{itemize}

\item Hybrid measures. These measures try to maximize more than one criterion with a single expression that finds a good trade-off between the criteria used. The hybrid quality measures implemented are:
\begin{itemize}
    \item \emph{Unusualness:} It is defined as the weigthed relative accuracy of a rule and tries to maximize generality and realiability:
    \begin{equation}
    WRAcc(R) = \frac{n(Cond)}{n_s}\left(\frac{n(Cond \wedge Target_{value})}{n(Cond)} - \frac{n(Target_{value})}{n_s}\right)\label{Unusualness}
    \end{equation} 

    \item \emph{True Positive Rate (TPR) or Sensitivity:} It measures the proportion of covered examples that has been correctly classified.
    \begin{equation}
      TPR(R) = \frac{n(Cond \wedge Target_{value})}{n(Target_{value})}
    \end{equation}

    \item \emph{False Positive Rate (FPR):} It measures the proportion of examples that are covered that do not belong to the target variable.
    \begin{equation}
        FPR(R) = \frac{n(Cond \wedge \overline{Target_{value}})}{n(\overline{Target_{value}})} 
    \end{equation}
    \noindent where $\overline{Target_{value}}$ means the negation of $Target_{value}$, i.e., the examples that not belong to the target class.
\end{itemize}   
</code></pre>

<p>\end{itemize}</p>

<p>A more extended classification and definition of quality measures for SD is available in \cite{kais} and \cite{atzmueller2015subgroup}.</p>

<h2>Evolutionary fuzzy systems</h2>

<p>Evolutionary fuzzy systems (EFS) are the union of two powerfull tools for aproximate reasoning: evolutionary algorithms and fuzzy logic.</p>

<p>In one side, evolutionary algorithms \cite{eiben2003introduction} are stochastic algorithms based on the natural evolution to solve complex optimization problems. It is based on a population of representations of possible solutions, called chromosomes. A basic evolutionary algorithm scheme is:</p>

<p>\begin{enumerate}
    \item Generate the initial population
    \item Evaluate the chromosomes of the population. This is the most important and expensive part of the GA. In the algorithms of this package, quality measures described above are used as evaluation function.
    \item Select the chromosomes which the genetic operators will be applied.
    \item Apply the genetic operators. The most used are:
        \begin{itemize}
            \item Crossover operator. Which takes two chromosomes and generates two descendants as a combination of the elements of the parents.
            \item Mutation operator. Which takes a chromosome and changes randomly a gene (a value of the possible solution).
        \end{itemize}
    \item Replace the population with the new generated chromomes.
    \item Go to step 2 until a stopping criteria is reached. Normally this criteria is a number of evaluations or generations.
\end{enumerate}</p>

<p>These algorithms performs a global stochastic search through a huge search space efficiently. However, it is possible that these algorithms can not find an optimal solution (global optimum), but a good one (a local optimum) that can solve the problem too. Theys are well suited for SD because the problem of finding subgroups can be formulated from the optimization point of view coding rules as a parameters structure that optimize some measures. Additionally, different kinds of rules exists in SD (with inequality, with intervals, fuzzy rules&hellip;). This can change dramatically the size of the search space and genetic algorithms can adapt these structures easily without performance degradation. Likewise, the selection of the genetic operators can make GA a great candidate to introduce expert knowledge \cite{Carmona2014overview}.</p>

<p>On the other side, fuzzy logic \cite{zadeh1975concept} is an extension of the traditional set theory. Its main objective is to modelate and deal with imprecise knowledge, which is nearer to the human reasoning. The main difference with traditional set theory is that belonging degree is not zero or one, but a real value in [0,1]. This possibility allows to define fuzzy limits and the chance of overlapping between fuzzy sets. </p>

<p>A fuzzy variable is a set of ling√ºistic labels, e.g. low, medium and high, which are defined by overlapped fuzzy sets. This information is nearer to human reasoning and it is possible to calculate with precission the value of each belonging degree. This expressivity allow to obtain simpler rules because continuous variables are more understandable for humans. A rule can be a set of fuzzy variables. To determine if the rule cover an example it is neccesary to calculate the belonging degree of each variable in the rule with respect to the example. If all variables has a belonging degree grater than zero, the example is covered.</p>

<p>Evolutionary fuzzy systems is the union of both techniques, and has three ways of work \cite{herrera2008genetic}: 
\begin{itemize}
    \item Evolutionary algorithms that evolve the fuzzy rules (changing the number of variables and their values) and uses fuzzy sets definitions defined by the user. This way of work is used by all the algorithms implemented in the SDEFSR package.
    \item Evolutionary algorithms that evolve the fuzzy sets, changing the number of fuzzy sets for each variable, its shapes, etc.
    \item Evolutionary algorithms that evolve both rules and fuzzy sets.
\end{itemize}</p>

<h1>The SDEFSR package</h1>

<p><strong>SDEFSR</strong> is a package entirely written on R from scratch. This package includes the most important EFS for SD presented throughout the literature up to the moment. In addition, <strong>SDEFSR</strong> has the capacity to read data in different standard and well-known formats such as ARFF (Weka), KEEL, CSV and <em>data.frame</em> objects. Similarly, all functions included in the <strong>SDEFSR</strong> package have default parameters with values recommended by the authors. This allows the algorithms to be executed in an easy way, without the necessity of knowing the parameters for final users.</p>

<h2>Algorithms of the package</h2>

<p>Now, we describe the algorithms that are available in our package. This contains three algorithms: SDIGA \cite{sdiga}, MESDIF \cite{mesdif} and NMEEF-SD \cite{nmeef}. In chapter \ref{uso} we describe how to use the algorithms.</p>

<h3>SDIGA (Subgroup Discovery Iterative Genetic Algorithm)</h3>

<p>The algorithm SDIGA is a subgroup discovery algorithm that extract rules with two possible representations: one with conjunctions of attribute-value pairs (called canonical representation) or one with a disyuntive normal form (DNF) in the antecedent. It follows an iterative schema with a genetic algorithm in his core to extract those rules, this genetic algorithm only extract one rule, the best of the population, and after the extraction a local optimization could be performed in order to improve the generality of the rule. As the algorithm follows an iterative schema, the genetic algorithm is executed one time after another until a stopping criteria is reached: the rule obteined by the genetic algorithm and after the local improvement phase must cover at least one example not covered by precedent rules and this rule must have a minimum confidence (see Equation \ref{confidence}).<br/>
SDIGA can work with lost data (represented by  the maximum value of the variable + 1), categorical variables and numerical ones using fuzzy logic with the latter to improve the interpretability of the results.</p>

<h4>Components of genetic algorithm of SDIGA</h4>

<p>As we mentioned above, a genetic algorithm is the core of SDIGA. Such genetic algorithm is the responsible of extract one rule per execution and this rule it is the best of the population at the final of the evolutive process.</p>

<h5>Representation schema  \label{esquemas}</h5>

<p>Each chromosome in the population represents a possible rule but it only represents the antecedent part of the rule because the consecuent is prefixed. SDIGA can handle two types of representation as we mentioned above, canonical and DNF.<br/>
Canonical representation is formed by a numerical vector of a fixed length equal to the number of variables with possibles values in a range in \([0, max]\) where \(max\) is the number of possible values for categorical variables or the number of fuzzy sets defined for numerical variables. This \(max\) value represents the no participation of that variable in the rule.<br/>
DNF representation is formed by a binary vector, also with fixed length equal to the sum of all number of values/fuzzy sets of the variables. Here, a variable does not participate in a rule when all of his values are equal to zero or one.</p>

<h5>Crossover operator \label{cruce}</h5>

<p>SDIGA follows a pure stationary schema which only crosses the two best chromosomes in an iteration of the evolutionary process. This crossover is performed by a two-point cross operator.</p>

<h5>Mutation operator \label{mutacion}</h5>

<p>Mutation operator is applied over all the population of parents and  chromosomes generated by crossover. The mutation probability is applied at every gene. This operator can be applied in two ways (both with the same probability):
\begin{itemize}
\item Eliminate the variable: it puts the no participation value in that variable.
\item Put a random value: it puts a random value in that variable. The no participation value is also included.
\end{itemize}</p>

<h5>Fitness function</h5>

<p>The function to define the quality of a chromose in SDIGA is a weighted average of three of the quality measures described in section \ref{medidas}. So the functions to maximize is: \begin{equation} f(x) = \frac{Sop(x) \cdot  w<em>1 + Conf(x) \cdot w_2 + Obj3(x) \cdot  w_3}{\sum</em>{i=1}<sup>{3}</sup> w<em>i} \label{fitnessSDIGA} \end{equation}<br/>
where:
\begin{itemize}
 \item \(Sop(x)\) is the local support. This support is a modification of support (see Equation \ref{support}) and can be crisp or fuzzy: 
    \begin{itemize}
      \item Crisp Support: \begin{equation} Sop_n(x) = \frac{Ne<sup>+(R_i)}{Ne</sup></em>{NC}} \end{equation}
      \item Fuzzy Support: \begin{equation} Sop<em>d(x) = \frac{\sum</em>{k} APC(E<sup>k,</sup> R<em>i)}{Ne</em>{NC}} \end{equation}
      where \(Ne^+(R_i)\) is the number of examples covered by the rule and it is not covered by previous rules. \(Ne_{NC}\) is the number of examples of the target variable that it is not covered by any rule yet and \(APC\) is the \textit{Antecedent Part Compatibility} equation calculted only with new covered examples (see Equation \ref{APC}).
    \end{itemize}</p>

<pre><code>\item $Conf(x)$ is the confidence defined as in Equation \ref{confidence} for the crisp case, but for fuzzy case is defined as the ratio of the sum of $APC$ expresion of the correctly covered examples and the sum of $APC$ all examples covered by the rule: \begin{equation} Conf_d(x) = \frac{\sum_{E^{CC}} APC(E^k, R_i)}{\sum_{E^{C}} APC(E^k, R_i)} \end{equation} where $E^{CC}$ are the correctly covered examples and $E^C$ are the covered examples.

\item $Obj3(x)$ is other quality measure of section \ref{medidas}
\item $w_i$ is the weight of objective \textit{i}
</code></pre>

<p>\end{itemize}</p>

<p>As this rules uses fuzzy logic, we need an expresion to determine when an example is covered or not by a rule and also determine the belonging degree of that example to the rule. This function is determined by the expression \(APC\) (<em>Antecedent Part Compatibility</em>) and it is calculate by the expresssion: \begin{equation} APC(e, R_i) = T(TC(\mu _{1}<sup>{1}(e_1),</sup> &hellip; , \mu _{n}<sup>{1}(e_i)),</sup> &hellip; , TC(\mu _{1}<sup>{i}(e_1),</sup> &hellip; , \mu _{n}<sup>{i}(e_i)))</sup> \label{APC} \end{equation} where:  </p>

<p>\begin{itemize}
  \item \(e_i\) is the value of the example for the variable \textit{i}
  \item \(\mu_{n}^{i}\) is the belonging degree to the set \textit{n} of the variable \textit{i}
  \item \(TC\) is the fuzzy t-conorm. In this case is the maximum t-conorm.
  \item \(T\) is the fuzzy t-norm. In this case is the minimum t-norm.
\end{itemize}</p>

<p>\(\mu_{n}^{i}\) will be one or zero if the variable is categorical and its value is the same of the rule or not. In case of numerical variable, \(\mu_{n}^{i}\) will be calculated following the triangular belonging degree function.
\begin{equation} 
\mu_{a,b,c} = \left{\begin{matrix}
0 &amp; x \leq a\ 
\frac{x-a}{b-a} &amp;  a \leq x \leq b\ 
 \frac{c-x}{c-b}&amp;  b \leq  x \leq  c\ 
 0 &amp;   c \leq  x
\end{matrix}\right.
\end{equation}</p>

<h5>Replace operator</h5>

<p>To get next population for the next iteration a replace operator must be performed. This operator sort the population by fitness value and keep only the <em>n</em> best chromosomes, being <em>n</em> the size of the population.</p>

<h4>Local optimization</h4>

<p>After the genetic algorithm returns a rule, this rule could be improved by means of a local optimization based on a <em>hill climbing first best</em> local search. The algorithm eliminate one by one a variable and if the rule has a support and confidence greater than or equal the actual, the process starts again with that variable eliminated.</p>

<h3>MESDIF (Multiobjetive Evolutionary Subgroup DIscovery Fuzzy rules)</h3>

<p>MESDIF is a multi-objective genetic algorithm that extract fuzzy rules. The representation used could be canonical or DNF (see Section \ref{esquemas}). This algorithm follows the SPEA2 \cite{spea} approach where an elite population is used along the whole evolutive process. At the end of the process, the rules stored in the elite population where returned as result.<br/>
The multi-objective approach is based on a niches schema based on the dominance in the Pareto front. This schema allows to find non-dominated individuals to fill the elite population, that has a fixed size. In Figure \ref{MESDIFEsquema} we see a basic schema of the algorithm.</p>

<p>\begin{figure}[htbp]
  \centering 
  \includegraphics{MESDIFSchema.PNG}
  \caption{MESDIF operations schema}
  \label{MESDIFEsquema}
\end{figure}</p>

<h4>Components of the genetic algorithm.</h4>

<p>Now we describe briefly the components of the genetic algorithm to show how it works.\pagebreak</p>

<h5>Initial Population generation \label{biasedInit}</h5>

<p>The initial population operator performed by MESDIF produce random chromosomes in two ways: one percentage of chromosomes are produced ramdomly and the rest are produced also randomly, but with the difference that only a maximum number of variables could participate in the rule. This produce more generality in the generated rules.</p>

<h5>Fitness function</h5>

<p>To obtain the fitness value of each chromosome, MESDIF follows this steps: 
\begin{itemize}</p>

<p>\item First, we look for dominated and non-dominated individuals in both populations. We call \textit{strength of an individual} the number of individuals that this domain.
\item An initial fitness value \(A_i\) is calculated for each individual, such value is the sum of the strength of all dominators of individual \textit{i}. So, we have to minimize this value and for non-dominated individuals is zero.
\item Due to this system could fail if there are a lot of non-dominated individuals, additional information about density is included. This density is computed by the nearest neighbour approach and it is calculate by \begin{equation} D(i) = \frac{1}{\sigma<sup>k</sup> + 2} \end{equation} where \(\sigma^k\) is the k-th nearest neighbour.
\item The final adaptation value is the sum of initial fitness and density information \begin{equation} Fitness(i) = D(i) + A(i) \label{fitnessMESDIF} \end{equation}</p>

<p>\end{itemize}</p>

<h5>Truncation operator</h5>

<p>As the elite population has a fixed size, we need a truncation operator to truncate the elite population if all non-dominated individuals can not fit in elite population. To make this truncation, the operator take all non-dominated individuals and calculate the distance among every one. Then, the two closest individuals are taken and eliminate the individual with his k-th nearest neighbour with a minor distance. This process is repeated until non-dominated indivuals fit in elite population.</p>

<h5>Fill operator</h5>

<p>If the number of non-dominated individuals are less than the size of the elite population, we need to fill elite population with dominated individuals. The operator sort the individuals by its fitness value and copy the <em>n</em> first individuals to elite population, where <em>n</em> is the size of the elite population.</p>

<h5>Genetic operators</h5>

<p>The genetic operators of MESDIF are:
\begin{itemize}
  \item A two-point crossover operator (see Section \ref{cruce})
  \item A biased mutation operator, the functionality is the same operator of SDIGA (see Section \ref{mutacion}) but it is applied over a population of selected individuals.
  \item A selection operator based in a binary tournament selection. This selection is only applied over the individuals of elite population.
  \item A replacement of the selected population based on the direct replace of the \textit{k} worse individuals of the population. \textit{k} is the number of individuals returned after crosses and mutations.
\end{itemize}</p>

<h3>NMEEF-SD (Non-dominated Multi-objective Evolutionary algorithm for Extracting Fuzzy rules in Subgroup Discovery)</h3>

<p>NMEEF-SD is another multi-objectve genetic algorithm based in the NSGA-II \cite{nsga} approach. This algorithm has a fast sorting algorithm and a reinitialisation based on coverage if the population does not evolve for a period.<br/>
This algorithm only works with a canonical representation (see Section \ref{esquemas}). In \cite{estudioNMEEF} a study is presented where it reflects the low quality of rules obtained with a DNF representation.</p>

<h4>Evolutionary process</h4>

<p>The evolutionary process follows this steps
\begin{itemize}
  \item An initial biased population \(P_t\) is generated (see Section \ref{biasedInit}).
  \item The genetic operators are applied over \(P_t\) obtaining \(Q_t\) with the same size as \(P_t\).
  \item \(P_t\) and \(Q_t\) are joined obtaining \(R_t\) and the fast sorting algorithm is applied over \(R_t\). The individuals are sorted forming different fronts in the following way: &ldquo;The first front (\(F_1\)) is composed of non-dominated individuals, the second front (\(F_2\)) is composed by individuales dominated by one individual; the third front (\(F_3\)) is composed by individuals dominated by two, and so on.&rdquo;
  \item After that, the algorithm generates the population of the next generation (\(P_{t+1}\)). First, the algorithm checks if the Pareto front covers new examples as it can be show in Figure \ref{procesoNMEEF}. If this condition is not satisfied during a period of the evolutionary process, a reinitialisation based on coverage is performed. Otherwise, the algorithm gets the next population (\(P_{t+1}\)) introducing, in order, the first complete fronts of \(R_t\). If the last front does not fit completely in \(P_{t+1}\) then, the front is sorted by the \textit{crowding distance} and first individuals are copied until \(P_{t+1}\) is filled.
  \item At the final of the evolutionary process the individuals in the Pareto front are returned.
\end{itemize}</p>

<p>\begin{figure}[hbtp]
  \centering
  \includegraphics{NMEEFSchema.PNG}
  \caption{Operation schema of NMEEF-SD}
  \label{procesoNMEEF}
\end{figure}</p>

<h4>Genetic operators</h4>

<p>The genetic operators of NMEEF-SD are a two-point crossover operator (see Section \ref{cruce}) and a biased mutation operator (see Section \ref{mutacion}).</p>

<h4>Reinitialisation operator</h4>

<p>This operator is applied if the Pareto front does not cover any new example during a 5% of the total number of evaluations. Then, the algorithm copy the non duplicated individuals in the Pareto front to \(P_{t+1}\) and the rest of individuals are generated by means of trying to cover one example of the target class with a maximum number of variables.</p>

<h3>FuGePSD</h3>

<p>FuGePSD \cite{FuGePSD} is another genetic algorithm that finds interesting subgroups of a given class. It uses a programming genetic approach, where individuals are represented as trees with variable length instead of vectors. Also, the consecuent of the rule is represented. This schema has the advantage of get rules for all possible values of a given target variable in one execution. Furthermore, FuGePSD has a variable population length, which can change over the evolutive process based on an competitive-cooperative approach, the Token Competition.</p>

<h4>Evolutionary process</h4>

<p>The evolutionary process of FuGePSD works as follow:</p>

<p>\begin{enumerate}
  \item Create a random population of a fixed length.
  \item Create a new population called Offspring Propulation, which is generated via genetic operators.
  \item Join original population and offspring and execute the token competition procedure
  \item Get global fitnees and replace best population if necessary.
  \item return to 2 until number of generations is reached.
  \item Apply to the best population a Screening function and return the resulting rules.
\end{enumerate}</p>

<p>The key function on this scheme is the token competition procedure, which promove diversity on the population, and some of the genetic operator will bring specifity to the rules.</p>

<h4>Genetic operators</h4>

<p>The genetic operators used by FuGePSD are:
\begin{itemize}
  \item Crossover: it takes to parents and to random subtrees of each one. The crossover will cross the subtrees if the grammar are correct.
  \item Mutation: Change randomly a variable of an indiviual with a random value.
  \item Insertion: Inserts a new node on an indivudual, This node is a variable with a random value.
  \item Dropping: Remove a subtree of an individual.
\end{itemize}</p>

<p>This genetics operators will be applied with a probability given by the user.</p>

<h4>Token Competition procedure</h4>

<p>The token competition is the key procedure of FuGePSD, this brings diversity on the population keeping the best individuals. The algorithm works as follows: let a token be an example of the datasets, each individual can catch a token if the rule can cover it. If its occur, a flag is setted at that token and this token cannot be catchet by other rules.</p>

<p>So, the token competition works as follows:</p>

<p>\begin{enumerate}
  \item Sort the population in descending order by their individual fitness value.
  \item In order, each individual takes as much token as it can. This action is storied in a value for each individual called penalized fitness:
  \begin{equation}
  PenalizedFitness(R_i) = nusualness(R_i) * \frac{count(R_i)}{ideal(R_i)}
  \end{equation}
  Where \(count(R_i)\) is the number of tokens the rule really seized and \(ideal(R_i)\) is the maximum number of tokens the rule can seize.
  \item Remove individuals with \(PenalizedFitness = 0\).
\end{enumerate}</p>

<h4>Screening Function</h4>

<p>At the end of the evolutive process, the screening function is launched to get the users quality rules only. This rules must reach a minimum level of confidence and sensitivity. The function has an external parameter (ALL_CLASS) which if true, the algorithm will return, at least, one rule per value of the target variable, or at least the best rule of the population if false.</p>

<h2>Package architecture</h2>

<p>The main advantage of the <strong>SDEFSR</strong> package is that it provides all evolutionary fuzzy systems for SD that exists in the bibliography. Algorithms included are an important kind of SD methods that are not availble in R at the moment. Therefore, this package provides to the R community a brand new possibility for data mining and data analysis.</p>

<p>The base of the package is defined by two S3 classes. This classes are:
\begin{itemize}
    \item \emph{SDEFSR_Dataset}. This object defines a dataset and contains information about it. Such information are stored in the following fields:
    \begin{itemize}
        \item \emph{relation}. Defines the name of the relation that this dataset belongs to, e.g. &ldquo;german credit&rdquo;. 
        \item \emph{attributeNames}. Store the names of the attributes.
        \item \emph{attributeTypes}. A character that defines the data type of the attribute, i.e. &#39;r&#39; for real values, &#39;e&#39; for integers and &#39;c&#39; for categoricals variables.
         \item \emph{min}. A vector with the minimum value for numerical variabes, for categorical variables, this value is zero.
         \item \emph{max}. A vector with the maximum value for numerical variables or the number of different categorical values that has a categorical variable.
         \item \emph{nVars}. The number of variables in the dataset (excluding the target class).
         \item \emph{data}. A list that contains each example of the dataset. The categorical variables in data are codified. This means that a categorical value is represented as a value in [0, max -1] that represents the position of the value. For example, in the \emph{german-credit} dataset, the variable \emph{ForeignWorker} has two values: &ldquo;A201&rdquo; and &ldquo;A202&rdquo;, these values are represented in the \emph{data} field of an \emph{SDEFSR_Dataset} class as 0 or 1, respectively.
         \item \emph{class_names}. A vector with the categorical values of the target class. By default, the target variable is the last one. If it is not categorical, the methods that reads datasets fail. The user can select other target class when executing the algorithms.
         \item \emph{examplesPerClass}. A list with the number of examples that belongs to each class.
         \item \emph{lostData}. A logical that indicate the presence of lost data.
         \item \emph{covered}. A logical vector with length the number of examples that indicates which examples ar covered by the generated rules. This value by default is \emph{NA}, but it is used in some algorithms like SDIGA.
         \item \emph{fuzzySets}. A list that indicates the fuzzy sets definitions for each variable. This value is \emph{NA} by default and it is filled for each algorithm.
         \item \emph{crispSets}. A list that indicates the crisp sets obtained from the fuzzy sets. As this value is infered from \emph{fuzzySets}, this is NA too.
         \item \emph{sets}. A vector that defines the number of fuzzy sets that each variable has or the number of categorical values.
         \item \emph{categoricalValues}. A list that contains a vector of names of each categorical variable or \emph{NA} if the variable is numerical.
         \item \emph{Ns}. The number of examples in the dataset.
         \end{itemize}</p>

<pre><code>     This class also exports the well-known S3 methods \emph{print()} and \emph{summary()} that show the data structure without codification and a summary with basic information about the dataset respectively.

     \item \emph{SDEFSR\_Rules}. This class is a list that contains the rules generated by an algorithm. To know the number of rules generated, it is possible to use \emph{length(SDEFSR\_RulesObject)}. Each rule has the following fields:
     \begin{itemize}
        \item \emph{rule}. The string that represent the rule description.
        \item \emph{qualityMeasures}. A list that contains the quality measures of the rule. Such measures are the same as described in the quality measures section.
     \end{itemize}

     This object must be returned for all the SD algorithms of this package, and it is neccesary to make an analysis of the generated rules. This object is necessary to the exported functions \emph{plotRules()} that plots an FPR vs TPR graph that allows the visualization of rules, and \emph{orderRules()} that return other \emph{SDEFSR\_Rules} object with the rules sorted by a given quality measure in descendant order. Likewise, this object overloads the subset operator (\emph{&#39;[&#39;}) to allow filtering operations easily.
</code></pre>

<p>\end{itemize}</p>

<p>Additionally, the package has a general function that reads datasets in ARFF, KEEL or CSV format called \emph{read.dataset()} and \emph{SDEFSR_DatasetFromDataFrame()} to transform a \emph{data.frame} into a \emph{SDEFSR_Dataset}. </p>

<h1>Use of SDEFSR package \label{uso}</h1>

<p>In this section we are going to explain how to use this package. This package tries to use in a really simple way subgroup discovery algorithms and also without any dependencies.</p>

<h2>Installing and load the package.</h2>

<p>The package SDEFSR is now available at CRAN servers, so it can be installed as any other package by simply typing: </p>

<pre><code class="r">install.packages(&quot;SDEFSR&quot;)
</code></pre>

<p>Also, the develop version is available into GitHub at <a href="http://github.com/aklxao2/SDEFSR">http://github.com/aklxao2/SDEFSR</a>, feel free to clone and contribute if you wish. If you wish to use the development version you need to install the package <code>devtools</code> using the commando <code>install_github</code>:</p>

<pre><code class="r">devtools::install_github(&#39;aklxao2/SDEFSR&#39;)
</code></pre>

<p><strong>SDEFSR</strong> depends only on the package <strong>shiny</strong> \cite{shiny}. This package is neccesary to use the user interface in a local way and if the package is not installed, it asked to the user to be install when running the function <code>SDEFSRR_GUI()</code>. 
Once installed the package has to be loaded before use it. The package can be loaded through <code>library()</code> or <code>require()</code>. After loading the package there are six datasets available as <em>SDEFSR_Dataset</em>: <code>carTra</code>, <code>carTst</code>, <code>germanTra</code>, <code>germanTst</code>, <code>habermanTra</code> and <code>habermanTst</code> that corresponds to <code>car</code>, <code>german</code> and <code>haberman</code> training and test datasets.  Also, rules generated by the SDIGA algorithm with defaults parameters over the <code>haberman</code> dataset are loaded as <code>habermanRules</code> as an <em>SDEFSR_Rules</em> object.  </p>

<h2>Loading a dataset</h2>

<p>In order to use SD algorithms available in the <strong>SDEFSR</strong> package, a <code>SDEFSR_Dataset</code> object is neccesary. This object can be generated using the <em>read.dataset()</em> function. This function reads &ldquo;.dat&rdquo; files with the KEEL data mining tool format, ARFF files (&ldquo;.arff&rdquo;) from WEKA or even CSV files (&ldquo;.csv&rdquo;).
Assuming the files <code>iris.dat</code>, <code>iris.arff</code> and <code>iris.csv</code> corresponding to the
classical iris dataset in KEEL, ARFF and CSV formats respectively in the working directory, the loading of these files will be as follows: </p>

<pre><code class="r">irisFromKEEL &lt;- read.dataset(&quot;iris.dat&quot;) 
irisFromARFF &lt;- read.dataset(&quot;iris.arff&quot;)
irisFromCSV &lt;- read.dataset(&quot;iris.csv&quot;)
</code></pre>

<p>Note that the function detects the type of data by the extension. To read csv files, the function has optional parameters that defines the separator used between fields, the decimal separator, the quote indicator and the <code>NA</code> identifier as parameters. By default, this options and values are <code>sep = &quot;,&quot;, quote = &quot;\&quot;&quot;, dec = &quot;.&quot;</code> and <code>na.strings = &quot;?&quot;</code> respectively. It is important to remark that sparse ARFF data is not supported.</p>

<p>If the dataset is not available in this formats, it is possible to obtain a <em>SDEFSR_Dataset</em> object from a <code>data.frame</code>. This <code>data.frame</code> could be loaded by <code>read.table()</code> or similar functions. Eventually, the resulting <code>data.frame</code> has to be given to the <em>SDEFSR_DatasetFromDataFrame()</em> function. As we can see, this function allows the creation of datasets on the fly, as in this example: </p>

<pre><code class="r">library(SDEFSR)
df &lt;- data.frame(matrix(data = runif(1000), ncol = 10))
#Add class attribute (It must be the last attribute and it must be categorical)
df[,11] &lt;- c(&quot;Class 0&quot;, &quot;Class 1&quot;, &quot;Class 2&quot;, &quot;Class 3&quot;)
SDEFSR_DatasetObject &lt;- SDEFSR_DatasetFromDataFrame(df, relation = &quot;random&quot;)
#Load from iris dataset
irisFromDataFrame &lt;- SDEFSR_DatasetFromDataFrame(iris, &quot;iris&quot;)
</code></pre>

<p>This will assign to <em>SDEFSR_DatasetObject</em> a new dataset created randomly with 100 examples and 11 attributes. Note that the target variable must be categorical, because the SD algorithms can only deal with categorical target variables.\</p>

<p>The <code>SDEFSR\_DatasetFromDataFrame()</code> function has three additional parameters: <code>names</code>, <code>types</code>, and <code>classNames</code>. These allow the manual assignment of attribute names, their types, and a vector with values of target variable, respectively. Leaving the default values (<code>NA</code>), the function automatically retrieves these values through the information found on the dataset. However, if the information in the dataset is not accurate, it could cause unexpected results for the SD algorithms.</p>

<h2>Obtaining information from a loaded dataset</h2>

<p>Once the dataset is loaded, it is possible to view a simple summary of its content by using the usual <code>summary()</code> function:</p>

<pre><code class="r">summary(irisFromDataFrame)
</code></pre>

<pre><code>## Summary of the SDEFSR_Dataset object: &#39;irisFromDataFrame&#39;
##  - relation: iris
##  - nVars: 4
##  - Ns: 150
##  - attributeNames: Sepal.Length, Sepal.Width, Petal.Length, Petal.Width, Species
##  - class_names: setosa, versicolor, virginica
##  - examplesPerClass: 50, 50, 50
</code></pre>

<p>Any of these values can be obtained individually using the <code>$</code> operator on the <code>SDEFSR_Dataset</code> object:</p>

<pre><code class="r">irisFromDataFrame$nVars
</code></pre>

<pre><code>## [1] 4
</code></pre>

<pre><code class="r">irisFromDataFrame$attributeNames
</code></pre>

<pre><code>## [1] &quot;Sepal.Length&quot; &quot;Sepal.Width&quot;  &quot;Petal.Length&quot; &quot;Petal.Width&quot; 
## [5] &quot;Species&quot;
</code></pre>

<p>Also, it is possible to print the dataset as a <code>data.frame</code> with the <code>print()</code> function.</p>

<h2>Executing Subgroup Discovery algorithms</h2>

<p>Once our datasets are ready to be used, it is time to execute one subgroup discovery algorithm. For example we want to execute the algorithm MESDIF. For the rest of the algorithm this steps are equal and only a few parameters are diferent, if you need help with the parameters, refer to the help of the function using <code>help(function)</code> or <code>?function</code>.<br/>
It is possible to execute an SD algorithm in two ways: Through a parameters file, specifying as argument the path to such file. Or by entering all the parameter names and values at the command line. You can find the structure of the parameters file, among other useful information, on the help pages of each function.</p>

<p>Assuming the <code>SDEFSR_Dataset</code> object <code>irisFromDataFrame</code> that has been loaded before, and that the <code>params.txt</code> parameters file is stored in the working directory, the easiest way to run the <code>MESDIF()</code> algorithm, for example, will be:
When we execute the algorithm, the results are shown in the console. This results are divided in three fields: </p>

<pre><code class="r">&gt; ruleSet &lt;- MESDIF(paramFile = &quot;param.txt&quot;)
</code></pre>

<pre><code class="r">ruleSet &lt;- MESDIF(paramFile = NULL, 
                  training = irisFromDataFrame, 
                  test = NULL,
                  output = c(&quot;optionsFile.txt&quot;, &quot;rulesFile.txt&quot;, &quot;testQM.txt&quot;),
                  seed = 0,
                  nLabels = 3,
                  nEval = 300,
                  popLength = 100,
                  eliteLength = 2,
                  crossProb = 0.6,
                  mutProb = 0.01,
                  RulesRep = &quot;can&quot;,
                  Obj1 = &quot;CSUP&quot;,
                  Obj2 = &quot;CCNF&quot;,
                  Obj3 = &quot;null&quot;,
                  Obj4 = &quot;null&quot;,
                  targetVariable = &quot;Species&quot;,
                  targetClass = &quot;virginica&quot;)
</code></pre>

<pre><code>## --------------------------------
## Algorithm: MESDIF
## Relation: iris
## Training dataset: training
## Test dataset: test
## Rules Representation:  CAN
## Number of evaluations: 300
## Number of fuzzy partitions: 3
## Population Length: 100
## Elite Population Length: 2
## Crossover Probability: 0.6
## Mutation Probability: 0.01
## Obj1: CSUP   (Weigth: )
## Obj2: CCNF   (Weigth: )
## Obj3: null   (Weigth: )
## Obj4: null
## Number of examples in training: 150
## Number of examples in test: 150
## Target Variable: Species
## Target Class: virginica
## --------------------------------
## 
##  
##  Searching rules for only one value of the target class... 
##  
## 
## GENERATED RULE 1
## Variable Sepal.Width = Label 1 ( 2 , 3.2 , 4.4 ) 
##  THEN virginica
## 
## 
## 
## GENERATED RULE 2
## Variable Petal.Width = Label 2 ( 1.3 , 2.5 , 3.7 ) 
##  THEN virginica
## 
## 
## 
## 
##  
##  Testing rules... 
##  
## 
## Rule 1 :
##   - N_vars: 2
##   - Coverage: 0.8
##   - Significance: 0.602743
##   - Unusualness: 0.02
##   - Accuracy: 0.357724
##   - CSupport: 0.286667
##   - FSupport: 0.245
##   - CConfidence: 0.358333
##   - FConfidence: 0.3528
##   - True Positive Rate: 0.86
##   - False Positive Rate: 0.77 
## 
## 
## 
## Rule 2 :
##   - N_vars: 2
##   - Coverage: 0.193333
##   - Significance: 27.673033
##   - Unusualness: 0.128889
##   - Accuracy: 0.9375
##   - CSupport: 0.193333
##   - FSupport: 0.201667
##   - CConfidence: 1
##   - FConfidence: 0.889706
##   - True Positive Rate: 0.58
##   - False Positive Rate: 0 
## 
## 
## 
## Global:
##   - N_rules: 2
##   - N_vars: 2
##   - Coverage: 0.496666
##   - Significance: 14.137888
##   - Unusualness: 0.074444
##   - Accuracy: 0.647612
##   - CSupport: 0.3
##   - FSupport: 0.223334
##   - FConfidence: 0.621253
##   - CConfidence: 0.679166
##   - True Positive Rate: 0.72
##   - False Positive Rate: 0.385
</code></pre>

<p>The output has three clearly defined sections:
\begin{itemize}
    \item The first one provides the user with information about the current execution, i.e., the values given to the parameters.
    \item After that, the rules obtained are shown one by one. These rules are numbered, starting at 1.
    \item Finally, the quality measures applied over the test (or training if <code>test = NULL</code>) dataset for each rule are shown. At the end of the results, the &ldquo;Global&rdquo; section shows quality measures as a summary, providing the mean of the quality measures of every rule.
\end{itemize}</p>

<p>As this output could be extremely large, the function also saves it to three files, one for each of the above sections. The name of these files by default are <em>optionsFile.txt</em>, <em>rulesFile.txt</em> and <em>testQM.txt</em> and being saved into the working directory, overwriting existing files. The format of these files is identical to the output generated by the algorithm, but divided into the sections described above.</p>

<h2>Analyzing the rules obtained</h2>

<p>After the execution of a SD algorithm, it returns a <code>SDEFSR_Rules</code> object that contains the rules obtained. Following the example, with the <code>ruleSet</code> object obtained we can plot a TPR vs FPR plot to view the reliability, and generality of the rule. Reliable rules has low values of FPR and great TPR values, and too much general variables has great values of both TPR and FPR. To plot the rules, we can use the function <code>plotRules()</code>:</p>

<pre><code class="text">library(ggplot2)
plotRules(ruleSet)
</code></pre>

<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAfgAAAH4CAIAAAApSmgoAAAACXBIWXMAAAsSAAALEgHS3X78AAAbvUlEQVR4nO3de3xT9d3A8e85SZukoQUkA6Rc5CIURHhAylVEBRGGslJwssoewSmiKApjAg5hTtAXuHnZgyjKRSfIUASGIGDloogIoiJTFBj3ghdoa6GlbW7n+YOqXCIU2ua0337ef5U2ye93SPrJ6S8nOYZlWQIA0Mu0ewIAgLJF6AFAOUIPAMoRegBQzmn3BCLIz88Ph8PFvLBpmqZpBoPBMp1SRA6HIxQKRX/c2NhYv98f/XHt2l6n0xkOh4v/kCgthmEYhhH9cXlIF4fX6y2jyahUHkNfUFBQ/EdbbGysx+PJz88v0ylF5PV67Rr3+PHj0Q+QXdsbHx8fCoUKCgqiPK7T6TRNM/rPqS6Xy+12V7aH9LFjxy7oCEBCf0FYugEA5Qg9AChH6AFAOUIPAMoRegBQjtADgHKEHgCUI/QAoByhBwDlCD0AKEfoAUA5Qg8AyhF6AFCO0AOAcoQeAJQj9ACgHKEHAOUIPQAoR+gBQDlCDwDKEXoAUI7QA4ByhB4AlCP0AKAcoQcA5Qg9AChH6AFAOUIPAMoRegBQjtADgHJOuycAnF/8ls/tnkJ5dKRlM7ungIqBPXoAUI7QA4ByhB4AlCP0AKAcoQcA5Qg9AChH6AFAOUIPAMoRegBQjtADgHKEHgCUI/QAoByhBwDlCD0AKEfoAUA5Qg8AyhF6AFCO0AOAcoQeAJQj9ACgHKEHAOUIPQAoR+gBQDlCDwDKEXoAUI7QA4ByhB4AlCP0AKAcoQcA5Qg9AChH6AFAOUIPAMoRegBQjtADgHKEHgCUI/QAoByhBwDlCD0AKEfoAUA5Qg8AyhF6AFCO0AOAcoQeAJQj9ACgHKEHAOUIPQAoR+gBQDlCDwDKEXoAUI7QA4ByhB4AlCP0AKAcoQcA5Qg9AChH6AFAOUIPAMoRegBQjtADgHKEHgCUI/QAoByhBwDlCD0AKEfoAUA5Qg8AyhF6AFCO0AOAcoQeAJQj9ACgHKEHAOUIPQAoR+gBQDlCDwDKEXoAUI7QA4ByhB4AlCP0AKAcoQcA5Qg9AChH6AFAOUIPAMoRegBQjtADgHKEHgCUI/QAoByhBwDlCD0AKEfoAUA5Qg8AyhF6AFCO0AOAcoQeAJRzXvQ1Q7vnjpq8JSExXoJ5mblVuw9/qH+S55SfW9krJ7+cMGZk55izr2vlfrHg2fnbCwL5cZ2Hjky53H3RswAAnMfFh17E8LQZNPH+dk4R/xcvjV7+2U1JnV3FuqKV+f6iXVeNfLTXJd8umTj7g+se7lHVKME8AADnUJLQ/8wwHS537NG3Ji2uPea+ZMe+BY+vaTm2n4iIWMe/fGPa3M/znTG1rrvr7usTnSIiMU1+fcuvahgiLldMOBg+eSNbt27ds2ePiHTp0sXr9RZ3A5xO0zTdbhv+KHA6nbaMKyIul8uyrCgPatf2OhyO6A9aIZT63WHjQ9rtdhf/IR0Khcp0MvqUJPRW/uevTfrLMgnkZHxfe9CUtjEbVp51mXDGqsVHuj8yqb3n6PLHZq5rO7ZHNUOMqk3bVRX/d5vnPffhZbc8VO3k7vy+ffs2b94sIh06dIiJibDgE5FpmoZhFP/ypcg0TVvGFZGYmJjoh96u7TVNXkmKrNTvDhsf0k5n6ex0IqKSLd20Tht/fztnOHfPkqdefu9wUtGNWac0KPzd4UN7d896brMh4mvYPlj0bSvz49lPL8rtcsdDvS6PL1q2SUlJSUlJEZHs7Ozjx48XcxKxsbEej6f4ly9FXq83Ly8v+uO6XK7c3NxwOBzlce3a3vj4+OgPWiGU+sPe3of0Be27VKlSpezmo09pPIuaVRq0vCy4NlPqhv3+kFgFGQczw1cW/ax24qX16gwZ3inuyMZF23wnd96t7LUzFztuGz+iuZe1eQAoY6W0Ru+NC+/bE7wxWWZNeXJLzRruhglFf/+ZiT1Ta0//2+TVEohLvn34yeECO7du/3Z/1mNfGyKOBn1GDetag+ADQBkxor/Ue17Z2dnFf7Hl5NJNTk5OmU4pIrv+zvX5fFlZWZVq6ca9cUv0xy3/jrRsVro3aONDOjMz84Ja5PP5ym4++vAyFwAoR+gBQDlCDwDKEXoAUI7QA4ByhB4AlCP0AKAcoQcA5Qg9AChH6AFAOUIPAMoRegBQjtADgHKEHgCUI/QAoByhBwDlCD0AKEfoAUA5Qg8AyhF6AFCO0AOAcoQeAJQj9ACgHKEHAOUIPQAoR+gBQDlCDwDKEXoAUI7QA4ByhB4AlCP0AKAcoQcA5Qg9AChH6AFAOUIPAMoRegBQjtADgHKEHgCUI/QAoByhBwDlCD0AKEfoAUA5Qg8AyhF6AFCO0AOAcoQeAJQj9ACgHKEHAOUIPQAoR+gBQDlCDwDKEXoAUI7QA4ByhB4AlCP0AKAcoQcA5Qg9AChH6AFAOUIPAMoRegBQjtADgHKEHgCUI/QAoByhBwDlCD0AKEfoAUA5Qg8AyhF6AFCO0AOAcoQeAJQj9ACgHKEHAOUIPQAoR+gBQDlCDwDKEXoAUI7QA4ByhB4AIvOnD2uctih7wYCEjk/uzHihV7sJW4MRLhb8dPwVMd6eMw6Ff7riByMbOz0p845HcbLnQugB4Fw8v/7HJ6/feZnjXJcxa1T56s2lh4tK79+0aHVedY9R6lMJHT+aXXgR1yP0ACqBwL7Ff+zRolGTFsl970vr2P3Z/fnLhjS+Y7lfRMS//I7GQ5YVSujAoge7Na1bt26dRp3v+dfeH3feg1un3vynf84a9Y+Pd756z4NzX7sracC8TEtECtbe36LPS4ctEXE06dsrd8nSk/v0/k2L1yf1vc5riIiVtfHptPZNGjRs0Sl13OK9ATl7FOuHD6b2a31Z/Xr1knpOePeoJVJ41twC60d2vHX0sKua/WbGjuDZt3k+hB6AeuHD8x54aGf/Jdv/+5+37jyxaVvENga/fH32NwOX7c7I2Luo56dPz9sR+vlnRp3BT41Ibvr7558Z1O+Wa79ctCLTkrx181e3SOt3qSEi4miaelP+kiUHwyc73zy1R4IhIrmrH7l3VZvpm/67e9MrKQfGD59z0H/mKIHDrz06u87f/7P/wPaZ3b5a+WG2FXETrO/Xfd369d0bxrYqOOs2wxGvcQrnxfynAUBFkr9hxScd71rQ1C1Su/fQ/g3HRbqQs9XoN6ft+HT1vGmb0lfsP3FDYeTieq4ZcN2D01ccvTlh/oZ2g570GRIUEXE26XdT4f3/PjB82MHF65v3ezjhbREJbn93zcG9NcYNfEdEgt8e/Spuuww9YxSjaou2cc9MvH/irr59bn/1b3XdIhFXZ2I6DLj1ck/E2wwOrRd7zu0n9ADUM0zTNIuWzA2Hw3H66nkw/4TfEuvYunG9R3/ZMe3Wnn2G9tswPPeXbsx99YDrRzw771X3p13/8HzVn77taJJ6c/DuJTtb7V/fIvXReONtERG3p0rzO+es/NPlDhEJBwOWmX/WKFWunfLxpu2r3lj41qO9x1l3L1l2X6Mz5yYiYsR544yIt+mIOd/2s3QDQD13m05N3p81d7dfrKNr5izeHRQxXK7QgT3fBCV0eOnC9fki4X3vpYdSp0wddVu3X+3+bGdB4OwDbEKhkIiIu8uAHl89+sgXNwzqFnfKTx2N+vW13hj71Hst+nWPP/lU4mzWq3vuwtmbciwp3DvnlisHL8nee+YowU8eu/bmOVaPoROee3l00hebdvrPntupzr7NnMh/epyC0ANQz2x017SJvtm9mif9z2/mxLRp6hSJ7fyHEZ5pnZKSez6wLXnQdYlVnM3TRiQtuvmKNu1vGLPnmiFN1z7x0s5TFr+Nqs3+x/n6kKHzD4fF1bFPN2/dfoM6uk4fpWFKX3PVR1ekdo//8Vuu9mNfTNs/smOzxkl95jeZOiWlRoszR5njGTis/sK+jRs2btL1xV/9deyNnrPndtpfIGfdZrXzHt5jWNZ5nwyiLTs7u+iJsxhiY2M9Hk9OTk6ZTikir9ebl5cX/XF9Pl9WVlY4fN4XYEqZXdsbHx/v3rgl+uOWf0daNivdG7TxIZ2ZmXlBLfL5fBc7WmjnlGvvcc9Nf6DBRe/nFm75c+cR8XPXj21+zmMuyw/W6AGg+Kyjrw/uMv7QoNmLK0rlhdADqGQcTUctX1gQd7G784bvt6/s+G2pzqjsEXoAlUxMQvXzHqeiCy/GAoByv7BHb2V/8q9ZSz6rnjKhS/baQNs+V17CMwLsc7xd64KCgigP6nQ6TdP0+/1RHtflcrndbluOL4BWEfudt2H8Tfe+dWTP2tV7Qyf+81z/ng9/YMMr8QCA0hBpj96/4Z9vNZ+4cdShtNHibjtyyfRdV7/2of/qG879HlsAKK8CgUB4/B8v6CquKf8oo8lEX8Q9eodTCgsDRce0WieysgpYywdQgZXDNwxFU6Q9+tiu9/zv46m9htfPOVj/72PffPPNrLuWdGV3HgAqpogvxsa2HL38nbbz33jvv8cdia1mbEhpV5PDMAGggjo94IFNz975zEenHWWwZ9fW9W/Gdnxw5gMdKtmRpwCgw+mhdzTsPuTOlmd/zIyjVsOK82ZfACgLgQ9Ht/rdmsSmNQwRiWk9bLjjL/em12paI/jDoexL/zBz/h/b7BjfLmW5r5nP8GdnZNccPGPBmE7x573ZKDg99GbNltd2bxn+9tN3PvnmlNyb8bHu0j/7IQBULEa13k+8/cKNJ1+yDG4e90zRP/PXj+ry1Nt3v9rYiO856e2ZfWJFCt57sMO0d8Z06m/zlEUk8hp9OOOjRQs/DoiIBPMOf/7+F9XumL3g6qYJtB4AIjAcDgmHrTO+FeeN+6UrRFmk0Dvb3fvinJ/+FT70z9tH7Mq1bozepACgPLJ+WDHu1zf8zRCJveaRN260flg1vm/vZwJZ+767ZPCsifGy0zr+7sS+vacZhd9/ta/x5A/KSzbPfzSNWbNZnYMzvg7K5SzTA6jUzli6WVHtxklLf/yniATFiO/x6NKZfWJD2VufGvTQa7tuG13K5wy4OJFCH9wy/d7nPi46TXow68u1W2o/dj+H3ABAMTmqt7ymlX/uIZFyG3qzbsfUAQ2KXow1zN+Pat25TR3eGgsAxWZUqxra9qnI9XZPRCTiqQTDBZnfHXPXrOktWqoJ52d89HlBh45NorV0w6kEz60SnkowEAjw6ZVRoPhUgn6/33pk9AVdRfFn3YT2zL0tyVenadO6da4a/trix29uVa+qu2rL2+fvLW54AQDly+mhD2x+fvJXv1nxbc4PR7dNyB//+3+3+PsHh/ILf9j1LB9dCQAV1Olr9OHDGYGOd3WoborU6tHzqkvjel/boAoH2wBARXbGi7GW5XA6Tr4xyul0mCbvkQKgQExMjPXEM3bPwjZnHXVjhQKFhYVhkcJA+KevDUdMrJPjbgBUUOFwOBgMFv/ypmmapp7mnRF6w/jmxd7VX/zp34tPfu0e8K+j8/u7ojozACgtoVDIvXHLBV3FurZLGU0m+k4Pvav//OOB+TZNpaKIf/KvP39txwQKRbx2jCs2ba+IxPe+1aaRcX5HWpaL9wThHPT8bQIAiIjQA4BynCIQAIpl7rBer+wu/GbbtsDlbet766T8ruYLE1dXiBOPsEcPAMUy6IWV6emLx1zd5p65q9JXvXJ3C2e13k+8nZ6+7uNPpjWZ99TbuSJGfM9Jb6enp7+35bP/a/ratHfsnnIRQg8AJVQRTzwCADg/RSceAQBEYlSUE4+wdAMApe3kiUf2H7J7HkUIPQCUvh9PPFIuRDjxiO3K+YlHTn1nLKLDzTtjy7GSvzM2OicecX348QVdRdNHILBHDwDKEXoAUI7QA4ByHF4JQD/TNE90aHtBly+7yUQfoQegn9PpdDorb+5UPWsBAM5G6AFAOUIPAMoRegBQjtADgHKEHgCUI/QAoByhBwDlCD0AKEfoAUA5Qg8AyhF6AFCO0AOAcoQeAJQj9ACgHKEHAOUIPQAoR+gBQDlCDwDKEXoAUI7QA4ByhB4AlCP0AKAcoQcA5Qg9AChH6AFAOUIPAMoRegBQjtADgHKEHgCUI/QAoByhBwDlCD0AKEfoAUA5Qg8AyhF6AFCO0AOAcoQeAJQj9ACgHKEHAOUIPQAoR+gBQDlCDwDKEXoAUI7QA4ByhB4AlCP0AKAcoQcA5Qg9AChH6AFAOUIPAMoRegBQjtADgHKEHgCUI/QAoByhBwDlCD0AKEfoAUA5Qg8AyhF6AFCO0AOAcoQeAJQj9ACgHKEHAOUIPQAoR+gBQDlCDwDKEXoAUI7QA4ByhB4AlCP0AKAcoQcA5Qg9AChH6AFAOUIPAMoRegBQznnR1wztnjtq8paExHgJ5mXmVu0+/KH+SZ5Tfm5lr5z8csKYkZ1jIl/fyvto2swTg0dcH29c9BwAAOd18aEXMTxtBk28v51TxP/FS6OXf3ZTUmdX8a5pHdsyZ/ILK3fXv3twCcYHABRDSUL/M8N0uNyxR9+atLj2mPuSHfsWPL6m5dh+IiJiHf/yjWlzP893xtS67q67r090iogYCVcNeXzqpZNnnHojX3/9dUZGhoi0bt3a7XYXdwOcTtM0Xa5iPsUAKGWl8tvncrksyyrmhUOhUMlHrFRKEnor//PXJv1lmQRyMr6vPWhK25gNK8+6TDhj1eIj3R+Z1N5zdPljM9e1HdujmiEiYhimaZ7+AsHWrVvff/99EWnevHm1atWKOQnTNE3TLP4TQ8mFozYSUBGUym/fBT1b+P3+ko9YqZRs6aZ12vj72znDuXuWPPXye4eTim7MOuWZOfzd4UN7d896brMh4mvYPniOmxs4cODAgQNFJDs7Oycnp5iTiI2N9Xg8xb98ycVHbSSgIij5b5/P5zt27Fjx9+hFxOv1lnDQSqU0lm7MKg1aXhZcmyl1w35/SKyCjIOZ4SuLflY78dJ6dYYM7xR3ZOOibb5qvPAKAFFWSmv03rjwvj3BG5Nl1pQnt9Ss4W6YUHSsjZnYM7X29L9NXi2BuOTbh5fOcACA4jMu6M+l6MjOzi7+iy02LN08+deojYWT3L1vtXsK+EVHWjYr4S34fL7MzMwLapHP5yvhoJUKb5gCAOUIPQAoR+gBQDlCDwDKEXoAUI7QA4ByhB4AlCP0AKAcoQcA5Qg9AChH6AFAOUIPAMoRegBQjtADgHKEHgCUI/QAoByhBwDlCD0AKEfoAUA5Qg8AyhF6AFCO0AOAcoQeAJQj9ACgHKEHAOUIPQAoR+gBQDlCDwDKEXoAUI7QA4ByhB4AlCP0AKAcoQcA5Qg9AChH6AFAOUIPAMoRegBQjtADgHKEHgCUI/QAoByhBwDlCD0AKEfoAUA5Qg8AyhF6AFCO0AOAcoQeAJQj9ACgHKEHAOUIPQAoR+gBQDlCDwDKEXoAUI7QA4ByhB4AlCP0AKAcoQcA5Qg9AChH6AFAOUIPAMoRegBQjtADgHKEHgCUI/QAoByhBwDlCD0AKEfoAUA5Qg8AyhF6AFCO0AOAcoQeAJQj9ACgHKEHAOUIPQAoR+gBQDlCDwDKEXoAUI7QA4ByhB4AlCP0AKAcoQcA5Qg9AChH6AFAOUIPAMoRegBQjtADgHKEHgCUI/QAoByhBwDlCD0AKOe0ewIVz/E/TTj5hdfrzcvLi/4EfD5fVlZWOByO8rh2bW98fPzxQKCgoCDK4zqdTtM0/X5/lMd1uVxutzsnJyfK44p9dzHKGnv0AKAcoQcA5Qg9AChH6AFAOUIPAMoRegBQjtADgHKEHgCUI/QAoByhBwDlCD0AKEfoAUA5Qg8AyhF6AFCO0AOAcoQeAJQj9ACgHKEHAOUIPQAoR+gBQDlCDwDKEXoAUI7QA4ByhB4AlCP0AKAcoQcA5Qg9AChH6AFAOUIPAMoRegBQrsKHvqCg4Pvvv7dl6HA4bMu4Bw4cCIVC0R/Xru3NzMzMy8uzZWjLsqI/aH5+/pEjR6I/rtj6kLZr6ErCsOWhXIrWrFmzYMGCGTNm2D2R6Gnfvv2qVauqV69u90Si5OGHH27btu2AAQPsnkiUrF69euHChc8//7zdE4medu3arVmzJiEhwe6JqFXh9+gBAOfmtHsCJZWYmNi1a1e7ZxFV/fr1c7lcds8iepKTkxs0aGD3LKInMTHx6quvtnsWUZWamhoTE2P3LDSr8Es3AIBzY+kGAJSr8Es3YuV9NG3micEjro837J5KmfEfWDl91kfHwjGXpd73+zZVDRER6/ulE8a9k3+JxzBrdR8xqleixufsSBuuUKW9f0+qDL/CdqvYobeObZkz+YWVu+vfPdjuqZQhK3fzwo317pqQWm3TM1NXHWj92wamiISPHI3r9dCjt9RXHICIG65Opb1/RSrLr7D9KvajyEi4asjjU4e00v0yTvjQnpzLml/qMLxJSc69ewpOfrfw6JHMjA9efu75+ev2ntD5OkvkDVen0t6/IpXlV9h+FTv0IoZhmmZF34jzsE7khd0eETE8HteJvDxLRMQq9NS74soOA26/2bd5+qvbVEYw4obrU2nv35Mqw6+w/Sr20k3lYMR5HYUFImIVFBR64uIMERGjevu0ISIi0q1r4rpdR8Ot66r7ZYm44fpU2vsX0cODp/wzExslHNhxJCz5u3aFGjb2iIhIaOeCKfO2F4qEjnzzwyW1q2uMYMQN16fS3r+IHvboyz+jSvv+yS/MnDJVzEtThtU3QzvmPfzu5Y/e0vndZx+Z4KkSe0m3IX29GkNw5obbPZ8yUmnvX0QPb5gCAOW07iUBAIoQegBQjtADgHKEHgCUI/QAoByhR3nnTx/WOG1Rod3TACouQg8AyhF62Mf6dlafhkOW5YmIBD/5c+vkx74ISejAoge7Na1bt26dRp3v+dfeYNFlC5cNaXzHcr+IiH/5HY2HLCsUK2vj02ntmzRo2KJT6rjFewO2bQdQzhF62Meo1Sf1ivX/Xp8vEvx80fKqtw5s7gh++frsbwYu252RsXdRz0+fnrcj9EvXzl39yL2r2kzf9N/dm15JOTB++JyD4WhOHqg4+AgE2Mio1Se1+ZNLPyzonvDmipq3LmnsELPV6Den7fh09bxpm9JX7D9xQ+EvvXM7uP3dNQf31hg38B0RCX579Ku47cGh9WKjOX2ggiD0sJNRq09q06eWbtgUl17vd2/VM8U6tu7h3qO/7Jh2a88+Q/ttGJ4b4UrB/BN+S8TtqdL8zjkr/3S5Q0TCwYDl4EPNgYhYuoGtjJq/7tf43b/++Z1GA/vWMkTC+95LD6VOmTrqtm6/2v3ZzoLAj2v0hssVOrDnm6CEDi9duD5fxNmsV/fchbM35VhSuHfOLVcOXpLDxzYBERF62Muo2adfo8++bpHWp4YhIo7maSOSFt18RZv2N4zZc82QpmufeGlnWEQktvMfRnimdUpK7vnAtuRB1yVWMVztx76Ytn9kx2aNk/rMbzJ1Sko1PuIRiIhPrwQA5dijBwDlCD0AKEfoAUA5Qg8AyhF6AFCO0AOAcoQeAJT7f5QRq1iI4I5dAAAAAElFTkSuQmCC" alt="plot of chunk unnamed-chunk-9"/>
Additionally, we can directly order the rule set by a quality measure with the function <code>orderRules</code> which returns another <code>SDEFSR_Rules</code> object with the sorted rules. By default, the ordering is done by confidence. </p>

<pre><code class="r">    rulesOrderedBySignificance &lt;- orderRules(ruleSet, by = &quot;Significance&quot;)
</code></pre>

<p>Filter rules by number of attribute-value pairs or keep those rules with a quality measure greater than a given threshold is an interesting functionality to keep only high-quality rules. Such filtering operations are quite simple to apply in <strong>SDEFSR</strong>. Using the subset operator (<code>&#39;[]&#39;</code>) and introducing the filtering conditions will generate a new <code>SDEFSR_Rules</code> object with the result:</p>

<pre><code class="r">#Apply filter by unusualness
filteredRules &lt;- ruleSet[Unusualness &gt; 0.05]
#We check only if the number of rules decrease.
#In this case, this value must be 1.
length(filteredRules)
</code></pre>

<pre><code>## [1] 1
</code></pre>

<pre><code class="r">#Also, you can make the filter as complex as you can
#Filter by Unusualness, TPr and number of variables:
filteredRules &lt;- ruleSet[Unusualness &gt; 0.05 &amp; TPr &gt; 0.9 &amp; nVars == 3]
#In this case, any of the rules match the conditions. Therefore, the
#number of rules must be 0.
length(filteredRules) 
</code></pre>

<pre><code>## [1] 0
</code></pre>

<h1>The user interface</h1>

<p>As we mentioned at the begin of this paper, the <strong>SDEFSR</strong> package provide to the user an interface to use the subgroup discovery task in a more easy way and also do some basic exploratory analysis tasks.<br/>
We can use the user interface by two ways: one of them is using the interface via web at: <a href="http://sdrinterface.shinyapps.io/shiny">http://sdrinterface.shinyapps.io/shiny</a> . This has the advantage of use the algorithm without having R installed in our system and also avoid expending process time in our machine. The other way is to use the interface is in a local way, having our local server. This could be possible simply using:</p>

<pre><code class="r">SDR_GUI()
</code></pre>

<p>This function launch the user interface in our predetermined web browser. As we can see in Figure \ref{pantallaInicial}. The page are organized into an options panel at the left and a tabbed panel at the right. Here is where our results are shown when we execute the algorithms.<br/>
The first we have to do is to load a KEEL dataset. If we want to execute a subgroup discovery algorithm we must load a training and a test file <strong>having the same</strong> <code>&#39;@relation&#39;</code> <strong>field in the KEEL dataset file.</strong><br/>
Once we select the dataset, automatically it shows a graph with information about the last variable defined in the dataset. The graph shows the distribution of examples having some values of the variable. At the rigth of this graphic we can see a table with some basic information, more extended if this variable is numerical.  </p>

<p>\begin{figure}[hbtp]
  \centering
  \includegraphics[width=\linewidth]{Interfaz1.PNG}
  \caption{Initial screen of the SDEFSR user interface.}
  \label{pantallaInicial}
\end{figure}
\pagebreak
When loaded a dataset we can do a lot of things, in Figure \ref{pantalla2} we can see all the possibilities we can do:
\begin{enumerate}
\item As we mentioned above, we can load a second dataset as a test (or training) file.
\item This lists have a double function. In one hand we could select the variable to visualize in the graph and in the other hand is to select the target variable for executing a subgroup discovery algorithm. Below the variable selection we can choose a target value of the variable to find subgroups about this value or search for all posible values of the target variable.
\item Here we can choose how the information will be visualized, for categorical variables we can choose show the information as a pie chart or as a historgram. For numerical variables, a histogram and a bloxpot are available.
\item Here we can choose the subgroup discovery algorithm  and his parameters, that it is shown below. Below the parameters, we have the button to execute the subgroup discovery algorithm.
\item This section allows the selection, for categorical variables, the values of that variable has that we can see in the graph. It is important to remark that it only &ldquo;hide&rdquo; the values in the graph, so it does not eliminate any example of the dataset.
\item It allows to visualize information about the training or test file.
\end{enumerate}</p>

<p>\begin{figure}[hbtp]
  \centering
  \includegraphics[width=\linewidth]{Interfaz2.PNG}
  \caption{Screen of the user interface once loaded a dataset.}
  \label{pantalla2}
\end{figure}</p>

<p>After the execution of a subgroup discovery algorithm, we go automatically to the tab <code>&#39;Rules generated&#39;</code>, this tab contains a table with all subgroups generated. If we want we could filter rules by variable, for example, typing into the <code>&#39;Search&#39;</code> field.<br/>
The tab <code>&#39;Execution Info&#39;</code> shows an echo of the parameters used for launch the algorithm. This echo is equal than the one we can see in R console.<br/>
The tab <code>&#39;Test Quality Measures&#39;</code> shows a table with quality measures of every rule applied to test dataset. The last row its a summary of results and shows the number of rules we have and the average results of every quality measure.</p>

<h1>Summary</h1>

<p>In this paper the <strong>SDEFSR</strong> package has been introduced. This package can use four subgroup discovery algorithms without any other dependencies to others tools/packages. Also, the posibility of read and load datasets in the KEEL dataset format is provided, but it can read dataset from ARFF format or load a data.frame object. Finally, a web-based interface is developed for make the work easier, even if we do not have R installed in our system.<br/>
The devolpment of the package will continue in the future, including more functionality to work with KEEL datasets, adding new subgroup discovery algorithms and also improve the web interface. As we can see we have a great job ahead, so we encourage other developers to participe adding tools or algorithms into the package, as we will do in futures releases.</p>

<p>\begin{thebibliography}{49}
\providecommand{\natexlab}[1]{#1}
\providecommand{\url}[1]{\texttt{#1}}
\expandafter\ifx\csname urlstyle\endcsname\relax
  \providecommand{\doi}[1]{doi: #1}\else
  \providecommand{\doi}{doi: \begingroup \urlstyle{rm}\Url}\fi</p>

<p>\bibitem[Alcal\&#39;{a}-Fdez et~al.(2011)]{keelDatasets}
J.~Alcal\&#39;{a}-Fdez, A.~Fernandez, J.~Luengo, J.~Derrac, S.~Garc\&#39;{i}a,
  L.~S\&#39;{a}nchez, and F.~Herrera.
\newblock Keel data-mining software tool: Data set repository, integration of
  algorithms and experimental analysis framework.
\newblock \emph{Journal of Multiple-Valued Logic and Soft Computing},
  2-3\penalty0 (17):\penalty0 255&ndash;287, 2011.</p>

<p>\bibitem[Atzmueller(2014)]{rsubgroup}
M.~Atzmueller.
\newblock \emph{rsubgroup: Subgroup Discovery and Analytics}, 2014.
\newblock URL \url{<a href="http://CRAN.R-project.org/package=rsubgroup">http://CRAN.R-project.org/package=rsubgroup</a>}.
\newblock R package version 0.6.</p>

<p>\bibitem[Atzmueller(2015)]{atzmueller2015subgroup}
M.~Atzmueller.
\newblock Subgroup discovery.
\newblock \emph{Wiley Interdisciplinary Reviews: Data Mining and Knowledge
  Discovery}, 5\penalty0 (1):\penalty0 35&ndash;49, 2015.</p>

<p>\bibitem[Atzmueller and Lemmerich(2012)]{atzmueller2012vikamine}
M.~Atzmueller and F.~Lemmerich.
\newblock Vikamine&ndash;open-source subgroup discovery, pattern mining, and
  analytics.
\newblock In \emph{Machine Learning and Knowledge Discovery in Databases},
  pages 842&ndash;845. Springer, 2012.</p>

<p>\bibitem[Atzmueller and Puppe(2006)]{atzmueller2006sd}
M.~Atzmueller and F.~Puppe.
\newblock Sd-map&ndash;a fast algorithm for exhaustive subgroup discovery.
\newblock In \emph{European Conference on Principles of Data Mining and
  Knowledge Discovery}, pages 6&ndash;17. Springer, 2006.</p>

<p>\bibitem[Atzmueller et~al.(2004)]{atzmueller2004towards}
M.~Atzmueller, F.~Puppe, and H.-P. Buscher.
\newblock Towards knowledge-intensive subgroup discovery.
\newblock In \emph{Proceedings of the Lernen - Wissensentdeckung - Adaptivit√§t</p>

<ul>
<li>Fachgruppe Maschinelles Lernen}, pages 111&ndash;117, 2004.</li>
</ul>

<p>\bibitem[Atzmueller et~al.(2016)]{Atzmueller2016965}
M.~Atzmueller, S.~Doerfel, and F.~Mitzlaff.
\newblock Description-oriented community detection using exhaustive subgroup
  discovery.
\newblock \emph{Information Sciences}, 329:\penalty0 965&ndash;984, 2016.</p>

<p>\bibitem[Back et~al.(1997)]{back1997handbook}
T.~Back, D.~B. Fogel, and Z.~Michalewicz.
\newblock \emph{Handbook of evolutionary computation}.
\newblock IOP Publishing Ltd., 1997.</p>

<p>\bibitem[Berlanga et~al.(2006)]{mesdif}
F.~Berlanga, M.~J. Del~Jesus, P.~Gonz{\&#39;a}lez, F.~Herrera, and M.~Mesonero.
\newblock Multiobjective evolutionary induction of subgroup discovery fuzzy
  rules: a case study in marketing.
\newblock \emph{Advances in Data Mining. Applications in Medicine, Web Mining,
  Marketing, Image and Signal Mining}, pages 337&ndash;349, 2006.</p>

<p>\bibitem[Carmona and Luengo(2015)]{Carmona2015387}
C.~J. Carmona and J.~Luengo.
\newblock A first approach in the class noise filtering approaches for fuzzy
  subgroup discovery.
\newblock \emph{Advances in Intelligent Systems and Computing}, 368:\penalty0
  387&ndash;399, 2015.</p>

<p>\bibitem[Carmona et~al.(2009)]{carmona2009analysis}
C.~J. Carmona, P.~Gonz{\&#39;a}lez, M.~J. del Jesus, and F.~Herrera.
\newblock An analysis of evolutionary algorithms with different types of fuzzy
  rules in subgroup discovery.
\newblock In \emph{Fuzzy Systems, 2009. FUZZ-IEEE 2009. IEEE International
  Conference on}, pages 1706&ndash;1711. IEEE, 2009.</p>

<p>\bibitem[Carmona et~al.(2010{\natexlab{a}})]{nmeef}
C.~J. Carmona, P.~Gonz{\&#39;a}lez, M.~J. del Jesus, and F.~Herrera.
\newblock Nmeef-sd: non-dominated multiobjective evolutionary algorithm for
  extracting fuzzy rules in subgroup discovery.
\newblock \emph{IEEE Transactions on Fuzzy Systems}, 18\penalty0 (5):\penalty0
  958&ndash;970, 2010{\natexlab{a}}.</p>

<p>\bibitem[Carmona et~al.(2010{\natexlab{b}})]{carmona2010evolutionary}
C.~J. Carmona, P.~Gonz{\&#39;a}lez, M.~J. del Jesus, C.~Romero, and S.~Ventura.
\newblock Evolutionary algorithms for subgroup discovery applied to e-learning
  data.
\newblock In \emph{Proceedings of Education Engineering Conference (EDUCON)},
  pages 983&ndash;990. IEEE, 2010{\natexlab{b}}.</p>

<p>\bibitem[Carmona et~al.(2012)]{OrOlive}
C.~J. Carmona, S.~Ramirez-Gallego, F.~Torres, E.~Bernal, M.~J. del Jesus, and
  S.~Garcia.
\newblock Web usage mining to improve the design of an e-commerce website:
  Orolivesur.com.
\newblock \emph{Expert Systems with Applications}, pages 11243&ndash;11249, 2012.</p>

<p>\bibitem[Carmona et~al.(2013)]{carmona2013fuzzy}
C.~J. Carmona, C.~Chrysostomou, H.~Seker, and M.~J. del Jesus.
\newblock Fuzzy rules for describing subgroups from influenza a virus using a
  multi-objective evolutionary algorithm.
\newblock \emph{Applied Soft Computing}, 13\penalty0 (8):\penalty0 3439&ndash;3448,
  2013.</p>

<p>\bibitem[Carmona et~al.(2014)]{Carmona2014overview}
C.~J. Carmona, P.~Gonz{\&#39;a}lez, M.~J. del Jesus, and F.~Herrera.
\newblock Overview on evolutionary subgroup discovery: analysis of the
  suitability and potential of the search performed by evolutionary algorithms.
\newblock \emph{WIREs Data Mining and Knowledge Discovery}, 4\penalty0
  (2):\penalty0 87&ndash;103, 2014.</p>

<p>\bibitem[Carmona et~al.(2015)]{carmona2015fuzzy}
C.~J. Carmona, V.~Ruiz-Rodado, M.~J. del Jesus, A.~Weber, M.~Grootveld,
  P.~Gonz{\&#39;a}lez, and D.~Elizondo.
\newblock A fuzzy genetic programming-based algorithm for subgroup discovery
  and the application to one problem of pathogenesis of acute sore throat
  conditions in humans.
\newblock \emph{Information Sciences}, 298:\penalty0 180&ndash;197, 2015.</p>

<p>\bibitem[Deb(2001)]{deb2001multi}
K.~Deb.
\newblock Multi-objective optimization using evolutionary algorithms wiley.
\newblock \emph{Chichester, UK}, 2001.</p>

<p>\bibitem[Deb et~al.(2000)]{nsga}
K.~Deb, S.~Agrawal, A.~Pratap, and T.~Meyarivan.
\newblock A fast elitist non-dominated sorting genetic algorithm for
  multi-objective optimization: Nsga-ii.
\newblock \emph{Lecture notes in computer science}, 1917:\penalty0 849&ndash;858,
  2000.</p>

<p>\bibitem[del Jesus et~al.(2007)]{deljesus2007evolutionary}
M.~J. del Jesus, P.~Gonz{\&#39;a}lez, F.~Herrera, and M.~Mesonero.
\newblock Evolutionary fuzzy rule induction process for subgroup discovery: a
  case study in marketing.
\newblock \emph{IEEE Transactions on Fuzzy Systems}, 15\penalty0 (4):\penalty0
  578&ndash;592, 2007.</p>

<p>\bibitem[Dem{\v{s}}ar et~al.(2013)]{demvsar2013orange}
J.~Dem{\v{s}}ar, T.~Curk, A.~Erjavec, {\v{C}}.~Gorup, T.~Ho{\v{c}}evar,
  M.~Milutinovi{\v{c}}, M.~Mo{\v{z}}ina, M.~Polajnar, M.~Toplak,
  A.~Stari{\v{c}}, et~al.
\newblock Orange: data mining toolbox in python.
\newblock \emph{The Journal of Machine Learning Research}, 14\penalty0
  (1):\penalty0 2349&ndash;2353, 2013.</p>

<p>\bibitem[Eiben and Smith(2003)]{eiben2003introduction}
A.~Eiben and J.~Smith.
\newblock Introduction to evolutionary algorithms, 2003.</p>

<p>\bibitem[Fogel(2006)]{fogel2006evolutionary}
D.~B. Fogel.
\newblock \emph{Evolutionary computation: toward a new philosophy of machine
  intelligence}, volume~1.
\newblock John Wiley &amp; Sons, 2006.</p>

<p>\bibitem[Freitas(2003)]{freitas2003survey}
A.~A. Freitas.
\newblock A survey of evolutionary algorithms for data mining and knowledge
  discovery.
\newblock In \emph{Advances in evolutionary computing}, pages 819&ndash;845.
  Springer, 2003.</p>

<p>\bibitem[Gamberger and Lavrac(2002)]{gamberger2002expert}
D.~Gamberger and N.~Lavrac.
\newblock Expert-guided subgroup discovery: Methodology and application.
\newblock \emph{Journal of Artificial Intelligence Research}, pages 501&ndash;527,
  2002.</p>

<p>\bibitem[Gamberger et~al.(2003)]{Gamberger200327}
D.~Gamberger, N.~Lavra\v{c}, and G.~Krsta\v{c}i\&#39;{c}.
\newblock Active subgroup mining: A case study in coronary heart disease risk
  group detection.
\newblock \emph{Artificial Intelligence in Medicine}, 28\penalty0 (1):\penalty0
  27&ndash;57, 2003.
\newblock \doi{10.1016/S0933-3657(03)00034-4}.
\newblock URL
  \url{<a href="https://www.scopus.com/inward/record.uri?eid=2-s2.0-0038500689&amp;partnerID=40&amp;md5=c145be9994977d3b34110e35c227cf31">https://www.scopus.com/inward/record.uri?eid=2-s2.0-0038500689&amp;partnerID=40&amp;md5=c145be9994977d3b34110e35c227cf31</a>}.
\newblock cited By 46.</p>

<p>\bibitem[Geng and Hamilton(2006)]{geng2006interestingness}
L.~Geng and H.~J. Hamilton.
\newblock Interestingness measures for data mining: A survey.
\newblock \emph{ACM Computing Surveys (CSUR)}, 38\penalty0 (3):\penalty0 9,
  2006.</p>

<p>\bibitem[Goldberg(1989)]{goldberg1989genetic}
D.~Goldberg.
\newblock Genetic algorithm in search, optimization and machine learing.
\newblock \emph{MA: Addison-Wesley Longman Publishing Co., Inc}, 1989.</p>

<p>\bibitem[Herrera(2008)]{herrera2008genetic}
F.~Herrera.
\newblock Genetic fuzzy systems: taxonomy, current research trends and
  prospects.
\newblock \emph{Evolutionary Intelligence}, 1\penalty0 (1):\penalty0 27&ndash;46,
  2008.</p>

<p>\bibitem[Herrera et~al.(2011)]{kais}
F.~Herrera, C.~J. Carmona, P.~Gonz{\&#39;a}lez, and M.~J. del Jesus.
\newblock An overview on subgroup discovery: foundations and applications.
\newblock \emph{Knowledge and information systems}, 29\penalty0 (3):\penalty0
  495&ndash;525, 2011.</p>

<p>\bibitem[H{\&ldquo;u}llermeier(2005)]{hullermeier2005fuzzy}
E.~H{\&quot;u}llermeier.
\newblock Fuzzy methods in machine learning and data mining: Status and
  prospects.
\newblock \emph{Fuzzy sets and Systems}, 156\penalty0 (3):\penalty0 387&ndash;406,
  2005.</p>

<p>\bibitem[Jin et~al.(2014)]{Jin2014SmartElectricity}
N.~Jin, P.~Flach, T.~Wilcox, R.~Sellman, J.~Thumim, and A.~Knobbe.
\newblock Subgroup discovery in smart electricity meter data.
\newblock \emph{IEEE Transactions on Industrial Informatics}, 10\penalty0
  (2):\penalty0 1327&ndash;1336, 2014.</p>

<p>\bibitem[John(1992)]{john1992holland}
H.~John.
\newblock Adaptation in natural and artificial systems, 1992.</p>

<p>\bibitem[Kav{\v{s}}ek and Lavra{\v{c}}(2006)]{kavvsek2006apriori}
B.~Kav{\v{s}}ek and N.~Lavra{\v{c}}.
\newblock Apriori-sd: Adapting association rule learning to subgroup discovery.
\newblock \emph{Applied Artificial Intelligence}, 20\penalty0 (7):\penalty0
  543&ndash;583, 2006.</p>

<p>\bibitem[Koza(1992)]{koza1992genetic}
J.~R. Koza.
\newblock \emph{Genetic programming: on the programming of computers by means
  of natural selection}, volume~1.
\newblock MIT press, 1992.</p>

<p>\bibitem[Lavra{\v{c}} et~al.(2004{\natexlab{a}})]{lavravc2004decision}
N.~Lavra{\v{c}}, B.~Cestnik, D.~Gamberger, and P.~Flach.
\newblock Decision support through subgroup discovery: three case studies and
  the lessons learned.
\newblock \emph{Machine Learning}, 57\penalty0 (1-2):\penalty0 115&ndash;143,
  2004{\natexlab{a}}.</p>

<p>\bibitem[Lavra{\v{c}} et~al.(2004{\natexlab{b}})]{lavravc2004subgroup}
N.~Lavra{\v{c}}, B.~Kav{\v{s}}ek, P.~Flach, and L.~Todorovski.
\newblock Subgroup discovery with cn2-sd.
\newblock \emph{The Journal of Machine Learning Research}, 5:\penalty0
  153&ndash;188, 2004{\natexlab{b}}.</p>

<p>\bibitem[Lemmerich et~al.(2011)]{Lemmerich2011129}
F.~Lemmerich, M.~Ifland, and F.~Puppe.
\newblock Identifying and presenting influence factors on student drop-outs by
  subgroup discovery.
\newblock pages 129&ndash;132, 2011.
\newblock URL
  \url{<a href="https://www.scopus.com/inward/record.uri?eid=2-s2.0-84874005031&amp;partnerID=40&amp;md5=35482d236cd80aad991d24fe681c313e">https://www.scopus.com/inward/record.uri?eid=2-s2.0-84874005031&amp;partnerID=40&amp;md5=35482d236cd80aad991d24fe681c313e</a>}.
\newblock cited By 0.</p>

<p>\bibitem[Leung et~al.(1992)]{leung1992rule}
K.~S. Leung, Y.~Leung, L.~So, and K.~F. Yam.
\newblock Rule learning in expert systems using genetic algorithm: 1, concepts.
\newblock In \emph{Proceedings of the 2nd International Conference on Fuzzy
  Logic and Neural Networks (IIZUKA 92)}, volume~1, pages 201&ndash;204, 1992.</p>

<p>\bibitem[{Lowerre}(1976)]{beamSearch}
B.~T. {Lowerre}.
\newblock \emph{{The Harpy speech recognition system}}.
\newblock PhD thesis, Carnegie-Mellon Univ., Pittsburgh, PA., 1976.</p>

<p>\bibitem[Meeng and Knobbe(2011)]{meeng2011flexible}
M.~Meeng and A.~Knobbe.
\newblock Flexible enrichment with cortana&ndash;software demo.
\newblock In \emph{Proceedings of BeneLearn}, pages 117&ndash;119, 2011.</p>

<p>\bibitem[Poitras et~al.(2016)]{Poitras2016204}
E.~Poitras, S.~Lajoie, T.~Doleck, and A.~Jarrell.
\newblock Subgroup discovery with user interaction data: An empirically guided
  approach to improving intelligent tutoring systems.
\newblock \emph{Educational Technology and Society}, 19\penalty0 (2):\penalty0
  204&ndash;214, 2016.
\newblock URL
  \url{<a href="https://www.scopus.com/inward/record.uri?eid=2-s2.0-84964491291&amp;partnerID=40&amp;md5=2c40f186d28100293654be5afba45ab4">https://www.scopus.com/inward/record.uri?eid=2-s2.0-84964491291&amp;partnerID=40&amp;md5=2c40f186d28100293654be5afba45ab4</a>}.
\newblock cited By 1.</p>

<p>\bibitem[Rodriguez et~al.(2013)]{Rodriguez20131810}
D.~Rodriguez, R.~Ruiz, J.~C. Riquelme, and R.~Harrison.
\newblock A study of subgroup discovery approaches for defect prediction.
\newblock \emph{Information and Software Technology}, 55\penalty0
  (10):\penalty0 1810 &ndash; 1822, 2013.</p>

<p>\bibitem[Schwefel(1995)]{schwefel1995evolution}
H.-P. Schwefel.
\newblock Sixth-generation computer technology series, 1995.</p>

<p>\bibitem[Stiglic and Kokol(2012)]{Stiglic2012348}
G.~Stiglic and P.~Kokol.
\newblock Discovering subgroups using descriptive models of adverse outcomes in
  medical care.
\newblock \emph{Methods of Information in Medicine}, 51\penalty0 (4):\penalty0
  348&ndash;352, 2012.
\newblock \doi{10.3414/ME11-02-0040}.
\newblock URL
  \url{<a href="https://www.scopus.com/inward/record.uri?eid=2-s2.0-84864857832&amp;partnerID=40&amp;md5=610901ec672dfe2b5c628c00a0476857">https://www.scopus.com/inward/record.uri?eid=2-s2.0-84864857832&amp;partnerID=40&amp;md5=610901ec672dfe2b5c628c00a0476857</a>}.
\newblock cited By 3.</p>

<p>\bibitem[Wickham and Chang(2015)]{hadley2015devtools}
H.~Wickham and W.~Chang.
\newblock \emph{devtools: Tools to Make Developing R Packages Easier}, 2015.
\newblock URL \url{<a href="http://CRAN.R-project.org/package=devtools">http://CRAN.R-project.org/package=devtools</a>}.
\newblock R package version 1.8.0.</p>

<p>\bibitem[Wrobel(2001)]{definicionSD}
S.~Wrobel.
\newblock Inductive logic programming for knowledge discovery in databases.
\newblock In \emph{Relational data mining}, pages 74&ndash;101. Springer, 2001.</p>

<p>\bibitem[Zadeh(1975)]{zadeh1975concept}
L.~A. Zadeh.
\newblock The concept of a linguistic variable and its application to
  approximate reasoning‚Äîi.
\newblock \emph{Information sciences}, 8\penalty0 (3):\penalty0 199&ndash;249, 1975.</p>

<p>\bibitem[Zitzler et~al.(2002)]{spea}
E.~Zitzler, M.~Laumanns, and L.~Thiele.
\newblock Spea2: Improving the strength pareto evolutionary algorithm for
  multiobjective optimization.
\newblock In \emph{International Congress on Evolutionary Methods for Design
  Optimization and Control with Applications to Industrial Problems}, pages
  95&ndash;100, 2002.</p>

<p>\end{thebibliography}</p>

</body>

</html>
